---
title: "1.3.Manual_Analysis_RSEM_Cases"
author: "Gerard Temprano Sagrera"
date: "2023-02-24"
output: html_document
---

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(data.table);library(edgeR);library(tidyverse);library(openxlsx);library(preprocessCore);library("org.Hs.eg.db");library(stats);library(qvalue);library(ggpubr)
```

### Load Annotation file to change format names:

```{r}
#ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v26.annotation.fixed.gtf.gz")
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v34.annotation.fixed.gtf.gz")
```

### Counts are already in TPM from RSEM:

```{r}
#counts.tpm <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.RSEM.txt")
counts.tpm <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.RSEM.V34.txt")

counts.tpm <- counts.tpm %>% remove_rownames %>% column_to_rownames(var="V1")

cases.tpm <- counts.tpm[,1:96]
```

### Subset significant genes when comparing cases vs controls:

```{r}
#sig.genes <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Significant_genes.txt", header = F)
#sig.genes <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Significant_genes_smoking.txt", header = F)

#cases.tpm <- cases.tpm[rownames(cases.tpm) %in% sig.genes$V1,]
```

### Create a case/control variable:

```{r}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_Complete.xlsx")
tec <- tec[tec$Seq_ID %in% colnames(cases.tpm),]
```

### Load Biological data to make comparissons:

```{r}
bio <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/20221115_Datos Pacientes.xlsx")
bio <- bio[bio$Muestra %in% colnames(cases.tpm),]

# Smoking:
bio$Smoking2 <- ifelse(bio$Smoking == 0, 0, 1)
```

### By Diameter:

```{r}
dge <- DGEList(counts = cases.tpm)

design <-
  model.matrix(
    ~ bio$aortc_diameter_mm + tec$FC + tec$Library + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit)

colnames(design) <-
  c("(Intercept)",
    "Diameter",
    "FC",
    "Library",
    "GC_Mean",
    "RIN",
    "DV200",
    "Qubit")
```

### Remove lowely expressed genes using LIMMA:

```{r}
keep <- filterByExpr(dge, design)

paste0("We keep ", sum(keep), " genes") # 10,551.
paste0("We remove ", sum(!keep), " genes") # 16,739.

dge <- dge[keep,,keep.lib.sizes=FALSE]
```

### Apply quantile normalization to counts:

```{r}
tpm.rows <- rownames(dge$counts)
tpm.cols <- colnames(dge$counts)

cases.tpm <- as.data.frame(normalize.quantiles(as.matrix(dge$counts)))

rownames(cases.tpm) <- tpm.rows
colnames(cases.tpm) <- tpm.cols
```

### Manual Differential Expression:

```{r}
print(paste0("There are ",sum(is.na(bio$aortc_diameter_mm))," NAs"))

Pvalue <-
  apply(cases.tpm, 1, function(x)
    summary(
      lm(
        x ~ bio$aortc_diameter_mm + tec$FC + tec$Library + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit
      )
    )[4][[1]][2, 4])
Coef <-
  apply(cases.tpm, 1, function(x)
    summary(
      lm(
        x ~ bio$aortc_diameter_mm + tec$FC + tec$Library + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit
      )
    )[4][[1]][2, 1])
```


### Plot P-values:

```{r}
#png(file = "C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Hist_Diameter.png", height = 10000, width = 10000, res = 1200)
#hist(Pvalue, main = "", cex.main = 2)
#dev.off()

#hist(Pvalue, main = "Diameter", cex.main = 2)
```

### Create qvalue object:

```{r}
qobj <- qvalue(p = Pvalue)
```

```{r, warning=FALSE}
#summary(qobj)
#hist(qobj)
#plot(qobj)
```

```{r}
Results.RSEM.V26 <- data.frame(
  pvalue = Pvalue,
  qvalue = qobj$qvalues,
  fdr = p.adjust(Pvalue, method = "fdr"),
  pi1 = 1-qobj$pi0,
  Beta = Coef
)

Results.RSEM.V26 <- merge(Results.RSEM.V26, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
Results.RSEM.V26 <- Results.RSEM.V26[,-1]
Results.RSEM.V26 <- Results.RSEM.V26[!duplicated(Results.RSEM.V26$gene_name),]
Results.RSEM.V26 <- Results.RSEM.V26 %>% remove_rownames %>% column_to_rownames(var="gene_name")

Results.RSEM.V26 <- Results.RSEM.V26[order(Results.RSEM.V26$pvalue),]
#write.xlsx(Results[order(Results$pvalue), ][1:10,],"C://Users/Gerard/Desktop/AAA/RNAseq/TopTable.xlsx", rowNames = TRUE)

nrow(Results.RSEM.V26[Results.RSEM.V26$fdr < 0.05,])

#fwrite(Results.RSEM.V26, "C://Users/Gerard/Desktop/AAA/RNAseq/DEG_Diameter/RSEM.V34.txt", sep = "\t", row.names = T)
```


### By Type:

```{r}
dge <- DGEList(counts = cases.tpm)

design <-
  model.matrix(
    ~ bio$TIPOANEURISMA + tec$Library + tec$GC_Mean + tec$RIN + tec$Qubit)

colnames(design) <-
  c("(Intercept)",
    "Type",
    "Library",
    "GC_Mean",
    "RIN",
    "Qubit")
```

### Remove lowely expressed genes using LIMMA:

```{r}
keep <- filterByExpr(dge, design)

paste0("We keep ", sum(keep), " genes") # 10,551
paste0("We remove ", sum(!keep), " genes") # 16,739

dge <- dge[keep,,keep.lib.sizes=FALSE]
```

### Apply quantile normalization to counts:

```{r}
tpm.rows <- rownames(dge$counts)
tpm.cols <- colnames(dge$counts)

cases.tpm <- as.data.frame(normalize.quantiles(as.matrix(dge$counts)))

rownames(cases.tpm) <- tpm.rows
colnames(cases.tpm) <- tpm.cols
```

```{r}
print(paste0("There are ", sum(is.na(
  bio$aortc_diameter_mm
)), " NAs"))

Pvalue <-
  apply(dge$counts, 1, function(x)
    summary(
      lm(
        x ~ bio$TIPOANEURISMA + tec$FC + tec$Library + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit
      )
    )[4][[1]][2, 4])
Coef <-
  apply(dge$counts, 1, function(x)
    summary(
      lm(
        x ~ bio$TIPOANEURISMA + tec$FC + tec$Library + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit
      )
    )[4][[1]][2, 1])
```


### Plot P-values:

```{r}
png(file = "C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Hist_Type.png", height = 10000, width = 10000, res = 1200)
hist(Pvalue, main = "", cex.main = 2)
dev.off()

hist(Pvalue, main = "Type", cex.main = 2)
```

### Create qvalue object:

```{r}
qobj <- qvalue(p = Pvalue)
```

```{r, warning=FALSE}
summary(qobj)
hist(qobj)
plot(qobj)
```

```{r}
Results <- data.frame(
  pvalue = Pvalue,
  qvalue = qobj$qvalues,
  fdr = p.adjust(Pvalue, method = "fdr"),
  pi1 = 1-qobj$pi0,
  Beta = Coef
)

Results <- merge(Results, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
Results <- Results[,-1]
Results <- Results[!duplicated(Results$gene_name),]
Results <- Results %>% remove_rownames %>% column_to_rownames(var="gene_name")

Results <- Results[order(Results$pvalue),]
#write.xlsx(Results[order(Results$pvalue), ][1:10,],"C://Users/Gerard/Desktop/AAA/RNAseq/TopTable.xlsx", rowNames = TRUE)

nrow(Results[Results$fdr < 0.05,])
```









