---
title: "Prepare Group Presentation"
author: "Gerard Temprano Sagrera"
date: 25/10/2022
output: "html_document"
---

```{r, warning=FALSE, message=FALSE, echo = FALSE}
library(openxlsx);library(data.table);library(stringr);
library(ggbiplot);
library(tidyr);library(data.table)
```


### First technical table, only with cases and controls:

```{r}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_Complete.xlsx")
nrow(tec) # 141 Samples.
table(tec$Case) # 44 Controls / 97 Cases
```

### Check all the samples have been sequenced:

```{r}
samples <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Correspondencias.xlsx")

mm <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/20221013_Formulario_TablaMuestras_Gerard.xlsx", startRow = 5)
mm <- mm[grep("HSP",mm$Nombre.de.la.muestra),]
mm2 <- mm[mm$Nombre.de.la.muestra%in%samples$EXHEUS_ID,]

length(intersect(samples$EXHEUS_ID, mm$Nombre.de.la.muestra))
```
### Combine 'multiQC' files:

```{r}
mqc1 <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/multiQC/multiqc_report_1.xlsx")
mqc2 <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/multiQC/multiqc_report_2.xlsx")

mqc <- rbind(mqc1, mqc2)

mqc <- mqc %>% fill(names(mqc), .direction = "up")

mqc <- mqc[-grep("_", mqc$Sample.Name),]
mqc <- mqc[order(mqc$Sample.Name),]

#write.xlsx(mqc, "C://Users/Gerard/Desktop/AAA/RNAseq/multiQC/multiqc_report.xlsx")
```


### Analyze number of reads numbers:

```{r}
ff <- list.files("/home/gerard/AAA/Results/", pattern = "*Log.final.out", recursive = T, full.names = T)
ff <- ff[ !grepl("_STARpass1", ff) ]

input_reads_df <- data.frame(File = character(), Input_Reads = numeric(), stringsAsFactors = FALSE)
mapped_reads_df <- data.frame(File = character(), Mapped_Reads = numeric(), stringsAsFactors = FALSE)

for (file in ff) {
  # Read the file as a character vector
  file_lines <- readLines(file)
  
  # Define regular expressions to match the parameter values
  input_reads_regex <- "Number of input reads \\|\\s*(\\d+)"
  mapped_reads_regex <- "Uniquely mapped reads number \\|\\s*(\\d+)"
  
  # Extract the parameter values using regular expressions
  input_reads <- round(as.numeric(sub(input_reads_regex, "\\1", file_lines[grep(input_reads_regex, file_lines)]))/1000000, 2)
  mapped_reads <- round(as.numeric(sub(mapped_reads_regex, "\\1", file_lines[grep(mapped_reads_regex, file_lines)]))/1000000, 2)
  
  # Extract the ID from the file path
  file_basename <- basename(file)
  file_id <- gsub("(^.*\\/)|([.].*$)", "", file_basename)
  
  # Append the extracted information to the corresponding data frames
  input_reads_df <- rbind(input_reads_df, data.frame(File = file_id, Input_Reads = input_reads, stringsAsFactors = FALSE))
  mapped_reads_df <- rbind(mapped_reads_df, data.frame(File = file_id, Mapped_Reads = mapped_reads, stringsAsFactors = FALSE))
}

ff <- list.files("/home/gerard/AAA/Results", pattern = ".metrics.tsv", recursive = T, full.names = T)

exonic_reads_df <- data.frame(Sample = character(), Exonic_Reads = numeric())

# loop through all files in ff
for (file_path in ff) {
  
  # read in the tsv file
  file_data <- fread(file_path, header = FALSE)
  
  # extract the Exonic Rate value from the second column where the first column equals "Exonic Rate"
  exonic_rate <- as.numeric(file_data[file_data$V1 == "Exonic Rate", 2])
  
  # extract the End 1 Antisense value from the second column where the first column equals "End 1 Antisense"
  end1_antisense <- as.numeric(file_data[file_data$V1 == "End 1 Antisense", 2])
  
  # calculate the Exonic_Reads parameter
  exonic_reads <- round((exonic_rate * end1_antisense)/1000000,2)
  
  # extract the Sample ID from the second column where the first column equals "Sample"
  sample_id <- file_data[file_data$V1 == "Sample", 2]
  
  # add the new row to the results dataframe
  exonic_reads_df <- rbind(exonic_reads_df, data.frame(Sample = sample_id, Exonic_Reads = exonic_reads, stringsAsFactors = FALSE))
}

write.csv(input_reads_df, file = "/home/gerard/AAA/Gene_Counts/input_reads.csv", row.names = FALSE)
write.csv(mapped_reads_df, file = "/home/gerard/AAA/Gene_Counts/mapped_reads.csv", row.names = FALSE)
write.csv(exonic_reads_df, file = "/home/gerard/AAA/Gene_Counts/exonic_reads.csv", row.names = FALSE)
```


```{r}
input_reads_df <- read.csv("C://Users/Gerard/Desktop/AAA/RNAseq/multiQC/input_reads.csv")
mapped_reads_df <- read.csv("C://Users/Gerard/Desktop/AAA/RNAseq/multiQC/mapped_reads.csv")
exonic_reads_df <- read.csv("C://Users/Gerard/Desktop/AAA/RNAseq/multiQC/exonic_reads.csv")

range(input_reads_df$Input_Reads, na.rm = T)
range(mapped_reads_df$Mapped_Reads, na.rm = T)
range(exonic_reads_df$Exonic_Reads, na.rm = T)

median(input_reads_df$Input_Reads, na.rm = T)
median(mapped_reads_df$Mapped_Reads, na.rm = T)
median(exonic_reads_df$Exonic_Reads, na.rm = T)
```


### Density plots:

```{r}
p <- density(input_reads_df$Input_Reads) # Input Reads
p2 <- density(mapped_reads_df$Mapped_Reads) # Mapped Reads
p3 <- density(exonic_reads_df$Exonic_Reads) # Exonic Reads 
```

```{r, include=TRUE}
plot(density(input_reads_df$Input_Reads), main = "Density Plot",ylim = range(p$y, p2$y, p3$y), xlim = range(p$x, p2$x, p3$x), col = "green")
lines(density(mapped_reads_df$Mapped_Reads), col = "blue")
lines(density(exonic_reads_df$Exonic_Reads), col = "red")
legend("topright",legend=c("Input Reads","Mapped Reads","Exonic Reads"), col = c("green","blue","red"), lty = 1, cex = 0.8, bty = "n", box.lty = 0)
```

### Ranges distribution:

```{r}
table(cut(mapped_reads_df$Mapped_Reads, breaks=c(0,5,10,20,30,max(mapped_reads_df$Mapped_Reads))))
```

### Exonic Reads:
```{r}
exonic_rate <- 100 * exonic_reads_df$Exonic_Reads/input_reads_df$Input_Reads

mean_exonic_rate <- mean(exonic_rate)



ggplot(data.frame(x = exonic_rate), aes(x)) +
  geom_histogram(binwidth = 1, fill = "lightblue", color = "blue") +
  ggtitle("Exonic Rate") +
  labs(x = "Exonic Reads/Total Reads", y = "Count") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5)) +
  geom_vline(xintercept=mean_exonic_rate, color="red", linetype="dashed", size=1)


  geom_text(aes(x=mean_exonic_rate-3, y=max(layer_data(p)$count), 
                  label=paste("Mean: ", round(mean_exonic_rate,2))),
                  color="red", hjust=0)
```

### Analyze annotation file:

```{r, message=FALSE, warning=FALSE}
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v34.annotation.fixed.gtf.gz", header = F, nThread = 6)
```

```{r, warning=FALSE}
merged <- readRDS("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/salmon.merged.gene_counts.rds")
#merged <- readRDS("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/salmon.merged.gene_counts_length_scaled.rds")
#merged <- readRDS("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/salmon.merged.transcript_counts.rds")

ff <- rownames(merged)

genes2 <- genes[genes$gene_id%in%ff,]

#write.xlsx(table(genes2$gene_type), "C://Users/Gerard/Desktop/AAA/RNAseq/Gene.Types.xlsx", col.names = F)
#write.xlsx(table(genes2$level), "C://Users/Gerard/Desktop/AAA/RNAseq/Annot.Levels.xlsx", col.names = F)

genes2 <- genes2[genes2$gene_type == "proteincoding" | genes2$gene_type == "lncRNA",]
genes2 <- genes2[genes2$level == 1 | genes2$level == 2,]
```


### Prepare Techincal Table 2:

```{r}
#corr <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Correspondencias.xlsx")

#fc <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/FlowCellType Date.xlsx", sheet = 1)
#s1 <- fc$S1
#sp <- fc$SP

#corr$FC <- ""
#corr[corr$Seq_ID%in%s1,]$FC <- "S1"
#corr[corr$Seq_ID%in%sp,]$FC <- "SP"


#dates <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/FlowCellType Date.xlsx", detectDates = T, sheet = 2)

#corr$Date <- ""
#dd1 <- dates$`2022-08-19`
#dd2 <- dates$`2022-08-29`

#corr[corr$Seq_ID%in%dd1,]$Date <- "2022-08-19"
#corr[corr$Seq_ID%in%dd2,]$Date <- "2022-08-29"


#plate <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/BatchPlate.xlsx")

#corr2 <- merge(corr, plate[,c("Seq_ID","Batch")], by = "Seq_ID")


#case <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/CaseControl.xlsx")

#corr3 <- merge(corr2, case[,c("SantPau_ID","Case")], by = "SantPau_ID")

# Add GC:
#samples <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Samples_Table.xlsx")
#samples <- samples[!is.na(samples$M.Reads),]
#samples$Sample.Name <- gsub("_R1","",samples$Sample.Name)
#colnames(samples)[colnames(samples) == '%.GC'] <- "GC_Mean"

#corr4 <- merge(corr3, samples[,c("Sample.Name","GC_Mean")], by.x = "SantPau_ID", by.y = "Sample.Name")


# Add RIN, DV200 and Qubit:
#rin <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/RIN.xlsx")

#corr_def <- merge(corr4, rin[,c("Sample.ID", "RIN", "DV200", "Qubit")], by.x = "SantPau_ID", by.y = "Sample.ID")


#write.xlsx(corr_def, "C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_CompleteV2.xlsx")

#names(corr_def)
```






