---
title: "RNAseq_AAA_Normalization"
author: "Gerard Temprano Sagrera"
date: "25/10/2022"
output: html_document
---
Load libraries
```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(edgeR);library(openxlsx);library(data.table);library(DESeq2);library(tximport);library(stringr);library(ggbiplot);library(dplyr)
```


Load and prepare annotation file

Filter genes: 
  1. Select "protein coding" and "lnRNA" genes.
  2. Select annotation levels "1" and "2".

```{r}
gtf <- fread("C://Users/Gerard/Downloads/gencode.v34.annotation.gtf.gz", header = F, nThread = 6)
gtf <- gtf[gtf$V3 == "gene",]
```

```{r, warning=FALSE}
genes <- gtf[,c("V4","V5","V9")]

genes$gene_id <- str_split_fixed(genes$V9,";",7)[,1]
genes$gene_id <- gsub("gene_id","",genes$gene_id)
genes$gene_id <- gsub("\\..*","",genes$gene_id)
genes$gene_id <- gsub("[^A-Za-z0-9]","",genes$gene_id)

genes$gene_type <- str_split_fixed(genes$V9,";",7)[,2]
genes$gene_type <- gsub("gene_type","",genes$gene_type)
genes$gene_type <- gsub("[^A-Za-z0-9]","",genes$gene_type)

genes$gene_name <- str_split_fixed(genes$V9,";",7)[,3]
genes$gene_name <- gsub("gene_name","",genes$gene_name)
genes$gene_name <- gsub("[^A-Za-z0-9]","",genes$gene_name)

genes$level <- str_split_fixed(genes$V9,";",7)[,4]
genes$level <- gsub("level","",genes$level)
genes$level <- gsub("[^A-Za-z0-9]","",genes$level)

genes$length <- genes$V5-genes$V4

genes <- genes[,-c("V4","V5","V9")]

merged <- readRDS("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/salmon.merged.gene_counts.rds")

ff <- rownames(merged)

# Select quantified genes:
genes2 <- genes[genes$gene_id%in%ff,]
# Select by gene type:
genes2 <- genes2[genes2$gene_type == "proteincoding" | genes2$gene_type == "lncRNA",]
# Select by annotation level:
genes2 <- genes2[genes2$level == 1 | genes2$level == 2,]
# Remove genes with the same ID:
genes2 <- genes2[!(duplicated(genes2$gene_id) | duplicated(genes2$gene_id, fromLast = TRUE)), ]
# Order by gene id:
genes2 <- genes2[order(genes2$gene_id),]
```


Save salmon counts and length of the genes
```{r}
counts <- assays(merged)$counts
# Select by annotation and type:
counts <- counts[rownames(counts)%in%genes2$gene_id,]
colnames(counts) <- gsub("_R1","",colnames(counts))
counts <- counts[order(rownames(counts)),]
dim(counts)

identical(genes2$gene_id,rownames(counts))

#fwrite(genes2[,c("gene_id","length")], "C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Lengths.txt", sep = "\t")
#fwrite(counts, "C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.salmon.txt", sep = "\t", row.names = T)
```


Apply TPM transformation to salmon counts
```{r}
counts <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.salmon.txt")
lengths <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Lengths.txt")
lengths <- lengths$length/1000
rnames <- counts$V1
counts <- counts[,-1]
rownames(counts) <- rnames

r_tpm <- function(dfr,len)
{
  dfr1 <- sweep(dfr,1,len,`/`)
  sc.f <- colSums(dfr1)/(10^6)
  return(sweep(dfr1,2,sc.f,`/`))
}

counts.tpm <- as.data.frame(r_tpm(counts, lengths))
```


#### Sex identification ###
Using the raw counts to increase the number of available genes.
```{r}
xist <- genes2[genes2$gene_name=="XIST",]$gene_id
counts.sex.f <- counts[rownames(counts)%in%xist,]

y.gene <- c("SRY","ZFY","RPS4Y1","AMELY","TBL1Y","PCDH11Y","TGIF2LY","TSPY1","TSPY2","AZFa","USP9Y","DDX3Y","UTY","TB4Y","AZFb","CYorf15","RPS4Y2","EIF1AY","KDM5D","XKRY","PRY","PRY2","HSFY1","HSFY2","RBMY1A1","AZFc","DAZ1","DAZ2","DAZ3","DAZ4","CDY1","CDY2","VCY1","VCY2")

y.list <- genes2[genes2$gene_name%in%y.gene,]$gene_id
counts.sex.m <- counts[rownames(counts)%in%y.list,]

# Combine the XIST gene with the mean expression of all Y chromosome genes:
counts.sex <- rbind(counts.sex.f, data.frame(t(colMeans(counts.sex.m))))
counts.sex <- as.data.frame(t(counts.sex))
colnames(counts.sex) <- c("XIST","Chromosome_Y")
```

```{r}
counts.sex[rownames(counts.sex) %in% "HSP075",]
```


Apply log transformation to the sex counts:
```{r}
counts.sex2 <- log(counts.sex)
counts.sex2[sapply(counts.sex2, is.infinite)] <- 0
```


Apply normalization to the sex counts:
```{r}
counts.sex2 <- as.data.frame(scale(counts.sex2))
colMeans(counts.sex2)
apply(counts.sex2, 2, sd)
counts.sex2$ID <- rownames(counts.sex2)
```


```{r}
ggplot(counts.sex2, aes(x = XIST, y = Chromosome_Y, label=ID)) +
  theme_bw() +
  geom_text(colour = "lightblue",
            size = 3) +
  geom_text(data = counts.sex2[rownames(counts.sex2) %in% c("HSP214","HSP217","HSP075"),],
             colour = "red",
             size = 3) +
  ggtitle("Gender RNAseq") +
  theme(plot.title = element_text(hjust = 0.5))

#ggsave("C://Users/Gerard/Downloads/Test2.png", height = 10)
```



### Check the sex between the controls we could not identify ###
```{r}
counts.sex[rownames(counts.sex) %in% c("HSP214", "HSP217"),]
```

```{r}
counts.sex$Sex.Gene <- ifelse(counts.sex$XIST > 300 | counts.sex$Chromosome_Y < 2, 2, 1 )
```


```{r}
sex <- read.xlsx("C://Users/Gerard/Downloads/20221115_Datos Pacientes.xlsx")
sex <- sex[2:nrow(sex),]

identical(rownames(counts.sex), sex$Muestra)
counts.sex <- cbind(counts.sex, sex$SEXO)
counts.sex[!counts.sex$Sex.Gene == counts.sex$`sex$SEXO`,]
```

```{r}
counts.sex %>%
  group_by(sex$SEXO) %>%
  summarise_at(vars(XIST, Chromosome_Y), list(name = range))
```




Use EdgeR to normalize the counts:
```{r, echo = FALSE}
#cc <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Techinc_1.xlsx")
#group <- cc[,c("Case/Control")]

# Check they have the same order:
#identical(colnames(counts),cc$Nombre.de.la.muestra)

#y <- DGEList(counts, group = group)
#dim(y)

#y2 <- calcNormFactors(y)
```


Use deseq2 to normalize the counts:
```{r, echo = FALSE}
#cc <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Techinc_1.xlsx")
#group <- cc[,c("Nombre.de.la.muestra","Case/Control")]
#colnames(group) <- c("ID","Status")
#rownames(group) <- group$ID
#group$Status <- factor(group$Status, levels = c("1","0"), labels = c("Case","Control"))
#group$Status <- relevel(group$Status, ref = "Control")

# Check they have the same order:
#identical(rownames(group),colnames(counts))
#all(rownames(group$ID) == colnames(counts))

#y <- DESeqDataSetFromMatrix(counts, colData = group, design = ~ Status)
#dim(y)

#keep <- rowSums(counts(y)) >= 10
#y <- y[keep,]
#counts(y)
```



Perform PCA, removing genes that do not have variability
```{r}
counts.tpm <- as.data.frame(r_tpm(counts, lengths))
```

```{r}
tec2 <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_CompleteV2.xlsx")
cc <- tec2[,c("Case")]
cc[cc == 1] <- "Case"
cc[cc == 0] <- "Control"
counts.tpm <- t(counts.tpm)

# Remove genes that have 0 variability:
counts_var <- counts.tpm[,apply(counts.tpm, 2, var, na.rm=TRUE) != 0]

my_pca <- prcomp(counts_var, center = T, scale. = T)
```


PCA Plots
```{r, include=TRUE}
plot(my_pca, main = "Variance Explained")
```


```{r, include=TRUE}
plot(cumsum(my_pca$sdev^2 / sum(my_pca$sdev^2)), type="b", ylab = "Cummulative proportion of Variance Explained")
```


PCA Plot
```{r, include=TRUE}
ggbiplot(my_pca, labels = rownames(counts_var), groups = cc, var.axes = F, obs.scale = 1, var.scale = 1) +
  scale_color_discrete(name = "") +
  theme(legend.direction = "horizontal", legend.position = "top")
```


Plot the histograms of the 9 smallest PC1
```{r}
par(mfrow = c(3, 3))
pca1.min <- head(order(predict(my_pca)[, 1]), 9)
for (i in 1:9) hist(as.numeric(counts_var[pca1.min[i], ]), main = paste0("Number ", i), xlab = "Gene expression")
```


Plot the histograms of the 9 largest PC1
```{r}
par(mfrow = c(3, 3))
pca1.max <- head(order(predict(my_pca)[, 1], decreasing = T), 9)
for (i in 1:9) hist(as.numeric(counts_var[pca1.max[i], ]), main = paste0("Number ", i), xlab = "Gene expression")
```




Technical batches names
```{r, echo = FALSE}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_CompleteV2.xlsx")
names(tec)
```


```{r}
table(tec$Batch, tec$Date)
```

```{r}
table(tec$Batch, tec$FC)
```

```{r}
table(tec$FC, tec$Date)
```


GC Content Histogram
```{r}
hist(tec$GC_Mean, 
     breaks = 40, 
     main = "GC_mean histogram", 
     xlab = "GC_mean")
```


Extreme Gc Content values
```{r}
# Left tail
par(mfrow = c(3, 3))
gc.min <- head(order(tec$GC_Mean), 9)
for (i in 1:9) hist(as.numeric(counts_var[gc.min[i], ]), main = paste0("Number ", i), xlab = "Gene expression")
```


```{r}
# Right tail
par(mfrow = c(3, 3))
gc.max <- head(order(tec$GC_Mean, decreasing = T), 9)
for (i in 1:9) hist(as.numeric(counts_var[gc.max[i], ]), main = paste0("Number ", i), xlab = "Gene expression")
```



# Covariates influence in the PCA: #

```{r}
df.predict <- data.frame(predict(my_pca))[, 1:2]
df.predict$SampleID <- rownames(df.predict)
df.all <- plyr::join(tec, df.predict, by = "SampleID", type = "right")
```


### Batch = Plate ###
```{r}
ggplot(df.all, aes(x = PC1, y = PC2)) + 
  theme_bw() + 
  coord_fixed() + 
  geom_point(aes(colour = as.factor(Batch), shape = as.factor(Case)), size = 2.5) +
  ggtitle("Technical variable: Batch") +
  theme(plot.title = element_text(hjust = 0.5))
```

### Batch = Plate ###
```{r}
ggplot(df.all, aes(x = PC1, y = PC2)) + 
  theme_bw() + 
  coord_fixed() + 
  geom_point(aes(colour = as.factor(Batch)), pch = 20, size = 3) + 
  ggtitle("Technical variable: Plate") +
  theme(plot.title = element_text(hjust = 0.5)) +
  labs(colour = "Plate")
```



### Flow Cell ###
```{r}
ggplot(df.all, aes(x = PC1, y = PC2)) + 
  theme_bw() + 
  coord_fixed() + 
  geom_point(aes(colour = FC, shape = as.factor(Case)), size = 2.5) +
  ggtitle("Technical variable: Flow Cell", )+
  theme(plot.title = element_text(hjust = 0.5))
```

### Flow Cell ###
```{r}
ggplot(df.all, aes(x = PC1, y = PC2)) + 
  theme_bw() + 
  coord_fixed() + 
  geom_point(aes(colour = FC), pch = 20, size = 3) + 
  ggtitle("Technical variable: Flow Cell", )+
  theme(plot.title = element_text(hjust = 0.5))
```


### Date ###
```{r}
ggplot(df.all, aes(x = PC1, y = PC2)) + 
  theme_bw() + 
  coord_fixed() + 
  geom_point(aes(colour = as.factor(Date)), pch = 20, size = 3) + 
  ggtitle("Technical variable: Date", )+
  theme(plot.title = element_text(hjust = 0.5))
```


### GC Content ###
```{r}
ggplot(df.all, aes(x = PC1, y = PC2)) + 
  theme_bw() + 
  coord_fixed() + 
  geom_point(aes(colour = as.factor(GC_Mean), shape = as.factor(Case)), size = 2.5) +
  ggtitle("Technical variable: GC Content", )+
  theme(plot.title = element_text(hjust = 0.5))
```


### RIN ###
```{r}
ggplot(df.all, aes(x = PC1, y = PC2)) + 
  theme_bw() + 
  coord_fixed() + 
  geom_point(aes(colour = RIN), pch = 20, size = 3) + 
  scale_colour_continuous(low = "lightblue", high = "darkblue") + 
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("Technical variable: RIN")
```


### DV200 ###
```{r}
ggplot(df.all, aes(x = PC1, y = PC2)) + 
  theme_bw() + 
  coord_fixed() + 
  geom_point(aes(colour = DV200), pch = 20, size = 3) + 
  scale_colour_continuous(low = "lightblue", high = "darkblue") + 
  ggtitle("Technical variable: DV200")
```


```{r}
ggplot(df.all, aes(x = PC1, y = PC2)) + 
  theme_bw() + 
  coord_fixed() + 
  geom_point(aes(colour = DV200, shape = as.factor(Case)), size = 2.5) +
  scale_colour_continuous(low = "lightblue", high = "darkblue") + 
  ggtitle("Technical variable: DV200")
```






### Qubit ###
```{r}
ggplot(df.all, aes(x = PC1, y = PC2)) + 
  theme_bw() + 
  coord_fixed() + 
  geom_point(aes(colour = Qubit), pch = 20, size = 3) + 
  scale_colour_continuous(low = "lightblue", high = "darkblue") + 
  ggtitle("Technical variable: Quvit")
```



























