---
title: "Limma Analysis Salmon"
author: "Gerard Temprano Sagrera"
date: "2022-12-09"
output: html_document
---

```{r, warning=FALSE, message=FALSE, echo = FALSE}
library(limma);library(data.table);library(edgeR);library(tidyverse);library(openxlsx)
```


```{r}
#limmaUsersGuide()
```

### Apply TPM transformation to salmon counts:

```{r}
# Load previously prepared salmon read counts:
counts <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.salmon.txt")
lengths <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Lengths.txt")
lengths <- lengths$length/1000
rnames <- counts$V1
counts <- counts[,-1]
rownames(counts) <- rnames

r_tpm <- function(dfr, len) {
  dfr1 <- sweep(dfr, 1, len, `/`)
  sc.f <- colSums(dfr1) / (10 ^ 6)
  return(sweep(dfr1, 2, sc.f, `/`))
}

counts.tpm <- as.data.frame(r_tpm(counts, lengths))
```

### Create a case/control variable:

```{r}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_CompleteV2.xlsx")
cc <- tec[,c("Case")]
cc[cc == 1] <- "Case"
cc[cc == 0] <- "Control"
table(tec$Case)
```

### Filter lowely expressed genes:

Keep genes that have more than 50% of non-zero counts across all samples:
```{r}
keep <- apply(counts.tpm, 1, function(x) sum(x == 0) <= 70)
sum(keep) # 21,355

counts.tpm <- counts.tpm[keep,]

# Check how this filter affects cases and controls:
cases.all.counts <- counts.tpm[,1:97]
controls.all.counts <- counts.tpm[,98:ncol(counts.tpm)]

sum(apply(cases.all.counts, 1, function(x) sum(x == 0) <= 48)) # 21,197
sum(apply(controls.all.counts, 1, function(x) sum(x == 0) <= 22)) # 20,499
```

### Convert TPM counts to DGElist object and remove rows with consistent 0 or low counts:

```{r}
dge <- DGEList(counts=cases.all.counts)

# Set "Control" to be the first level:
AAA <- factor(x = c(rep("Case",97), rep("Control",44)),
              levels=c("Control","Case"))

# Batch and FC are not included because are highly correlated with AAA:

design.all <- model.matrix(~AAA+tec$Date+tec$GC_Mean+tec$RIN+tec$FC+tec$Batch)
design.few <- model.matrix(~AAA+tec$Date+tec$GC_Mean+tec$RIN)

colnames(design.all) <- c("(Intercept)", "AAACase", "Date202220829", "GC_Mean", "RIN","FC","Plate")
colnames(design.few) <- c("(Intercept)", "AAACase", "Date202220829", "GC_Mean", "RIN")

keep.all <- filterByExpr(dge, design.all)
keep.few <- filterByExpr(dge, design.few)

dge.all <- dge[keep.all,,keep.lib.sizes=FALSE]
dge.few <- dge[keep.few,,keep.lib.sizes=FALSE]
```

### Apply TMM Normalization:

```{r}
dge.all <- calcNormFactors(dge.all, method = "TMM")
dge.few <- calcNormFactors(dge.few, method = "TMM")
```

### Apply voom transformation   :

```{r}
v.all <- voom(dge.all, design.all, plot = TRUE)
v.few <- voom(dge.few, design.few, plot = TRUE)
```

### Limma Differential Expression:

```{r}
fit.all <- lmFit(v.all, design.all)
fit.all <- eBayes(fit.all)
```

```{r}
fit.few <- lmFit(v.few, design.few)
fit.few <- eBayes(fit.few)
```

```{r}
lm.deg <- topTable(fit.all, coef=1, number = Inf)

summary(decideTests(fit.all))
```

```{r}
lm.deg <- topTable(fit.few, coef=1, number = Inf)

summary(decideTests(fit.few))
```

### Obtain Gene_ID from Ensembl_ID, using annotation file:

```{r}
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v34.annotation.fixed.gtf.gz")
lm.deg2 <- merge(lm.deg, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")

if(nrow(lm.deg) != nrow(lm.deg2)){print("Some genes have been lost")}
```












