---
title: "Manual Analysis Salmon Cases"
author: "Gerard Temprano Sagrera"
date: "2022-12-16"
output: html_document
---

```{r, warning=FALSE, message=FALSE, echo = FALSE}
library(data.table);library(edgeR);library(tidyverse);library(openxlsx);library(preprocessCore)
```

### Apply TPM transformation to salmon counts:

```{r}
# Load previously prepared salmon read counts:
counts <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.salmon.txt")
lengths <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Lengths.txt")
lengths <- lengths$length/1000
rnames <- counts$V1
counts <- counts[,-1]
rownames(counts) <- rnames

r_tpm <- function(dfr,len)
{
  dfr1 <- sweep(dfr,1,len,`/`)
  sc.f <- colSums(dfr1)/(10^6)
  return(sweep(dfr1,2,sc.f,`/`))
}

counts.tpm <- as.data.frame(r_tpm(counts, lengths))
```

### Create a case/control variable:

```{r}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_CompleteV2.xlsx")
cc <- tec[,c("Case")]
cc[cc == 1] <- "Case"
cc[cc == 0] <- "Control"
```

### Filter lowely expressed genes:

Keep genes that have more than 50% of non-zero counts across all samples:
```{r}
keep <- apply(counts.tpm, 1, function(x) sum(x == 0) <= 70)
sum(keep) # 21,355

counts.tpm <- counts.tpm[keep,]

# Check how this filter affects cases and controls:
cases.tpm <- counts.tpm[,1:97]
controls.tpm <- counts.tpm[,98:ncol(counts.tpm)]

sum(apply(cases.tpm, 1, function(x) sum(x == 0) <= 48)) # 21,197
sum(apply(controls.tpm, 1, function(x) sum(x == 0) <= 22)) # 20,499
```

### Apply quantile normalization to counts:

```{r}
tpm.rows <- rownames(cases.tpm)
tpm.cols <- colnames(cases.tpm)

cases.tpm <- as.data.frame(normalize.quantiles(as.matrix(cases.tpm)))

rownames(cases.tpm) <- tpm.rows
colnames(cases.tpm) <- tpm.cols
```


### Manual Differential Expression:

```{r}
Pvalue <- apply(counts.tpm, 1, function(x) summary(lm(x~tec$Case+tec$Date+tec$GC_Mean+tec$RIN))[4][[1]][2,4])

Pvalue[order(Pvalue)][1:10]

summary(lm(dge$counts[1,]~tec$Case+tec$Date+tec$GC_Mean+tec$RIN))
#summary(lm(dge$counts[1,]~tec$Case+tec$Date+tec$GC_Mean+tec$RIN))[4][[1]][2,4]
```

### Obtain Gene_ID from Ensembl_ID, using annotation file:

```{r}
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v34.annotation.fixed.gtf.gz")
lm.deg2 <- merge(lm.deg, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")

if(nrow(lm.deg) != nrow(lm.deg2)){print("Some genes have been lost")}
```















