---
title: "Limma Analysis Salmon Cases"
author: "Gerard Temprano Sagrera"
date: "2022-12-16"
output: html_document
---

```{r, warning=FALSE, message=FALSE, echo = FALSE}
library(data.table);library(edgeR);library(tidyverse);library(openxlsx);library(preprocessCore)
```

### Apply TPM transformation to salmon counts:

```{r}
# Load previously prepared salmon read counts:
counts <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.salmon.txt")
lengths <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Lengths.txt")
lengths <- lengths$length/1000
rnames <- counts$V1
counts <- counts[,-1]
rownames(counts) <- rnames

r_tpm <- function(dfr,len)
{
  dfr1 <- sweep(dfr,1,len,`/`)
  sc.f <- colSums(dfr1)/(10^6)
  return(sweep(dfr1,2,sc.f,`/`))
}

counts.tpm <- as.data.frame(r_tpm(counts, lengths))
```

### Select only cases on technical variables:

```{r}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_CompleteV2.xlsx")

# Select only cases:
tec <- tec[1:97,]
```

### Load Biological data to make comparissons:

```{r}
bio <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/20221115_Datos Pacientes.xlsx")
bio <- bio[2:nrow(bio),]

bio <- bio[bio$Status == "Case",]

# Create BMI:
bio$BMI <- bio$Weight_Kg/((bio$Height_cm/100)^2)
bio$BMI.di <- ifelse(bio$BMI > 25, 1, 0)
bio$BMI.tri <- ifelse(bio$BMI > 30, 2, ifelse(bio$BMI > 25, 1, 0))

# Transform age to dicotomic variable:
bio$age2 <- ifelse(bio$age <= 65, 0, 1)
table(bio$age2)

# Transform aortic diameter to dicotomic variable:
bio$aortc_diameter_mm2 <- ifelse(bio$aortc_diameter_mm > 55, 1, 0) 
table(bio$aortc_diameter_mm2, useNA = "always")
```


## Manual Analysis:


### Apply quantile normalization to  counts:

```{r}
tpm.rows <- rownames(counts.tpm)
tpm.cols <- colnames(counts.tpm)

counts.tpm <- as.data.frame(normalize.quantiles(as.matrix(counts.tpm)))

rownames(counts.tpm) <- tpm.rows
colnames(counts.tpm) <- tpm.cols
```

### Filter lowely expressed genes:

Keep genes that have more than 50% of non-zero counts across all samples:
```{r}
keep <- apply(counts.tpm, 1, function(x) sum(x == 0) <= 70)
#sum(keep) # 21,355

counts.tpm <- counts.tpm[keep,]

# Check how this filter affects cases and controls:
cases.tpm <- counts.tpm[,1:97]
controls.tpm <- counts.tpm[,98:ncol(counts.tpm)]

#sum(apply(cases.tpm, 1, function(x) sum(x == 0) <= 48)) # 21,197
#sum(apply(controls.tpm, 1, function(x) sum(x == 0) <= 22)) # 20,499
```

### Differential Expression by Dicotomic Age (0 <= 65 Years Old 1 > 65 Years Old):

```{r}
print(paste0("There are ", sum(is.na(bio$age2))," NAs"))
table(bio$age2, useNA = "always")

Pvalue <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$age2+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,4])
Coef <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$age2+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,1])
```

```{r}
Results.age <- as.data.frame(cbind(Pvalue, Coef))

Results.age[p.adjust(Pvalue, method = "fdr") < 0.05,]
```

### Differential Expression by Sex (1 Men 2 Women):

```{r}
print(paste0("There are ", sum(is.na(bio$SEXO))," NAs"))
table(bio$SEXO, useNA = "always")

Pvalue <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$SEXO+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,4])
Coef <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$TIPOANEURISMA+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,1])
```

```{r}
Results.sex <- as.data.frame(cbind(Pvalue, Coef))

Results.sex[p.adjust(Pvalue, method = "fdr") < 0.05,] # 9
```

### Differential Expression by Size of the Aneurym (1 > 55 mm 0 < 55 mm):

```{r}
print(paste0("There are ", sum(is.na(bio$aortc_diameter_mm2))," NAs"))
table(bio$aortc_diameter_mm2, useNA = "always")

Pvalue <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$aortc_diameter_mm2))[4][[1]][2,4])
Coef <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$aortc_diameter_mm2))[4][[1]][2,1])
```

```{r}
Results.size <- as.data.frame(cbind(Pvalue, Coef))

Results.size[p.adjust(Pvalue, method = "fdr") < 0.05,] # 0, even with no other variables.
```

### Differential Expression by Smoking status (0 Never 1 Past (+6 months) 2 Current):

Smoking1:
```{r}
print(paste0("There are ", sum(is.na(bio$Smoking))," NAs"))
table(bio$Smoking, useNA = "always")

Pvalue <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$Smoking+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,4])
Coef <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$Smoking+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,1])

summary(lm(dge$counts[1,]~bio$Smoking+tec$Date+tec$GC_Mean+tec$Batch))
```

```{r}
Results.smoke <- as.data.frame(cbind(Pvalue, Coef))

Results.smoke[p.adjust(Pvalue, method = "fdr") < 0.05,] # 0.
```


Smoking2:
```{r}
print(paste0("There are ", sum(is.na(bio$Smoking))," NAs"))
table(bio$Smoking, useNA = "always")

Pvalue <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$Smoking+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][3,4])
Coef <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$Smoking+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][3,1])
```

```{r}
Results.smoke <- as.data.frame(cbind(Pvalue, Coef))

Results.smoke[p.adjust(Pvalue, method = "fdr") < 0.05,] # 0.
```

### Differential Expression by Type of Aneurysm (1 Atherosclerotic 2 Inflammatory):

```{r}
print(paste0("There are ", sum(is.na(bio$TIPOANEURISMA))," NAs"))

Pvalue <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$TIPOANEURISMA+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,4])
Coef <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$TIPOANEURISMA+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,1])
```

```{r}
Results.type <- as.data.frame(cbind(Pvalue, Coef))

Results.type[p.adjust(Pvalue, method = "fdr") < 0.05,] # 0.
```

### Obtain Gene_ID from Ensembl_ID, using annotation file:

```{r}
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v34.annotation.fixed.gtf.gz")
lm.deg2 <- merge(lm.deg, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")

if(nrow(lm.deg) != nrow(lm.deg2)){print("Some genes have been lost")}
```



## Limma Analysis:

### Filter lowely expressed genes:

Keep genes that have more than 50% of non-zero counts across all samples:
```{r}
keep <- apply(counts.tpm, 1, function(x) sum(x == 0) <= 70)
sum(keep) # 21,355

counts.tpm <- counts.tpm[keep,]

# Check how this filter affects cases and controls:
cases.tpm <- counts.tpm[,1:97]
controls.tpm <- counts.tpm[,98:ncol(counts.tpm)]

sum(apply(cases.tpm, 1, function(x) sum(x == 0) <= 48)) # 21,197
sum(apply(controls.tpm, 1, function(x) sum(x == 0) <= 22)) # 20,499
```

### Convert TPM counts to DGElist object:

```{r}
dge <- DGEList(counts=cases.tpm)
```

### Apply TMM Normalization:

```{r}
dge <- calcNormFactors(dge, method = "TMM")
```


### DEG by Age:

```{r}
design <- model.matrix(~bio$age2+tec$Date+tec$GC_Mean+tec$Batch)

colnames(design) <- c("(Intercept)", "Age2", "Date202220829", "GC_Mean", "Plate")

v <- voom(dge, design, plot = TRUE)

# Limma Differential Expression:

fit <- lmFit(v, design)
fit <- eBayes(fit)

# Obtain differentially expressed genes between all conditions:

summary(decideTests(fit))
```

### DEG by Sex:

```{r}
design <- model.matrix(~bio$SEXO+tec$Date+tec$GC_Mean+tec$Batch)

colnames(design) <- c("(Intercept)", "Sex", "Date202220829", "GC_Mean", "Plate")

v <- voom(dge, design, plot = TRUE)

# Limma Differential Expression:

fit <- lmFit(v, design)
fit <- eBayes(fit)

# Obtain differentially expressed genes between all conditions:

summary(decideTests(fit))
```

### DEG by Aneurysm Size:

```{r}
keep <- !is.na(bio$aortc_diameter_mm2)

dge2 <- dge[,keep]

design <- model.matrix(~bio$aortc_diameter_mm2+tec$Date+tec$GC_Mean+tec$Batch)

colnames(design) <- c("(Intercept)", "Size", "Date202220829", "GC_Mean", "Plate")

v <- voom(dge2, design, plot = TRUE)

# Limma Differential Expression:

fit <- lmFit(v, design)
fit <- eBayes(fit)

# Obtain differentially expressed genes between all conditions:

summary(decideTests(fit))
```

### DEG by Smoking Status:

```{r}
keep <- !is.na(bio$Smoking)

dge2 <- dge[,keep]

design <- model.matrix(~bio$Smoking+tec$Date+tec$GC_Mean+tec$Batch)

colnames(design) <- c("(Intercept)","Smoking1","Smoking2","Date202220829", "GC_Mean", "Plate")

v <- voom(dge2, design, plot = TRUE)

# Limma Differential Expression:

fit <- lmFit(v, design)
fit <- eBayes(fit)

# Obtain differentially expressed genes between all conditions:

summary(decideTests(fit))
```

### DEG by Disease:

```{r}
table(bio$BMI.di, useNA = "always")
keep <- !is.na(bio$BMI.di)

dge2 <- dge[,keep]

design <- model.matrix(~bio$BMI.di+tec$Date+tec$GC_Mean+tec$Batch)

colnames(design) <- c("(Intercept)","Disease","Date202220829", "GC_Mean", "Plate")

v <- voom(dge2, design, plot = TRUE)

# Limma Differential Expression:

fit <- lmFit(v, design)
fit <- eBayes(fit)

# Obtain differentially expressed genes between all conditions:
summary(decideTests(fit))
```




















