---
title: "Limma Analysis Salmon"
author: "Gerard Temprano Sagrera"
date: "2022-12-09"
output: html_document
---

```{r, warning=FALSE, message=FALSE, echo = FALSE}
library(limma);library(data.table);library(edgeR);library(tidyverse);library(openxlsx)
```


```{r}
#limmaUsersGuide()
```

### Apply TPM transformation to salmon counts:

```{r}
# Load previously prepared salmon read counts:
counts <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.salmon.txt")
lengths <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Lengths.txt")
lengths <- lengths$length/1000
rnames <- counts$V1
counts <- counts[,-1]
rownames(counts) <- rnames

r_tpm <- function(dfr, len) {
  dfr1 <- sweep(dfr, 1, len, `/`)
  sc.f <- colSums(dfr1) / (10 ^ 6)
  return(sweep(dfr1, 2, sc.f, `/`))
}

counts.tpm <- as.data.frame(r_tpm(counts, lengths))
```

### Create a case/control variable:

```{r}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_CompleteV2.xlsx")
cc <- tec[,c("Case")]
cc[cc == 1] <- "Case"
cc[cc == 0] <- "Control"
table(tec$Case)
```

### Filter lowely expressed genes:

Keep genes that have more than 50% of non-zero counts across all samples:
```{r}
keep <- apply(counts.tpm, 1, function(x) sum(x == 0) <= 70)

sum(keep) # 21,355

counts.tpm <- counts.tpm[keep,]

# Check how this filter affects cases and controls:
cases.all.counts <- counts.tpm[,1:97]
controls.all.counts <- counts.tpm[,98:ncol(counts.tpm)]

#sum(apply(cases.all.counts, 1, function(x) sum(x == 0) <= 48)) # 21,197
#sum(apply(controls.all.counts, 1, function(x) sum(x == 0) <= 22)) # 20,499
```

### Convert TPM counts to DGElist object:

```{r}
dge <- DGEList(counts=counts.tpm)

# Set "Control" to be the first level:
AAA <- factor(x = c(rep("Case",97), rep("Control",44)),
              levels=c("Control","Case"))

# Batch and FC are not included because are highly correlated with AAA:

design <- model.matrix(~AAA+tec$Date+tec$GC_Mean+tec$RIN+tec$FC+tec$Batch)

colnames(design) <- c("(Intercept)", "AAACase", "Date202220829", "GC_Mean", "RIN","FC","Plate")
```

### Remove lowely expressed genes:

```{r}
keep <- filterByExpr(dge, design)
dge <- dge[keep,,keep.lib.sizes=FALSE]
```

### Apply TMM Normalization:

```{r}
dge <- calcNormFactors(dge, method = "TMM")
```

### Apply voom transformation:

```{r}
v <- voom(dge, design, plot = TRUE)
```

### Limma Differential Expression:

```{r}
fit <- lmFit(v, design)
fit <- eBayes(fit)
```

### Obtain differentially expressed genes between all conditions:

```{r}
summary(decideTests(fit))
```

### Obtain results in a table:

```{r}
lm.deg <- topTable(fit, coef="AAACase", number = Inf)
nrow(lm.deg)
```

### Obtain Gene_ID from Ensembl_ID, using annotation file:

```{r}
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v34.annotation.fixed.gtf.gz")
gene.names <- ann[ann$gene_id %in% rownames(lm.deg),]$gene_name

lm.deg2 <- merge(lm.deg, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")

if(nrow(lm.deg) != nrow(lm.deg2)){print("Some genes have been lost")}
```

### Volcano Plot:

```{r}
volcanoplot(fit, coef = "AAACase", highlight = 10, names = gene.names)
```






