---
title: "Limma Analysis Salmon Cases"
author: "Gerard Temprano Sagrera"
date: "2022-12-16"
output: html_document
---

```{r, warning=FALSE, message=FALSE, echo = FALSE}
library(data.table);library(edgeR);library(tidyverse);library(openxlsx);library(preprocessCore);library(qvalue)
```

### Apply TPM transformation to salmon counts:

```{r}
# Load previously prepared salmon read counts:
counts <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.salmon.txt")
lengths <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Lengths.txt")
lengths <- lengths$length/1000
rnames <- counts$V1
counts <- counts[,-1]
rownames(counts) <- rnames

r_tpm <- function(dfr,len)
{
  dfr1 <- sweep(dfr,1,len,`/`)
  sc.f <- colSums(dfr1)/(10^6)
  return(sweep(dfr1,2,sc.f,`/`))
}

counts.tpm <- as.data.frame(r_tpm(counts, lengths))
```

### Select only cases on technical variables:

```{r}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_CompleteV2.xlsx")

# Select only cases:
tec <- tec[1:97,]
```

### Load Biological data to make comparissons:

```{r}
bio <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/20221115_Datos Pacientes.xlsx")
bio <- bio[2:nrow(bio),]

bio <- bio[bio$Status == "Case",]

# Create BMI:
bio$BMI <- bio$Weight_Kg/((bio$Height_cm/100)^2)
bio$BMI.di <- ifelse(bio$BMI > 25, 1, 0)
bio$BMI.tri <- ifelse(bio$BMI > 30, 2, ifelse(bio$BMI > 25, 1, 0))
```

### Load Annotation file to change format names:

```{r}
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v34.annotation.fixed.gtf.gz")
```

## Manual Analysis:

### Apply quantile normalization to  counts:

```{r}
tpm.rows <- rownames(counts.tpm)
tpm.cols <- colnames(counts.tpm)

counts.tpm <- as.data.frame(normalize.quantiles(as.matrix(counts.tpm)))

rownames(counts.tpm) <- tpm.rows
colnames(counts.tpm) <- tpm.cols
```

### Filter lowely expressed genes manually:

Remove genes that have more than 50% of zero counts across all samples:
```{r}
keep <- apply(counts.tpm, 1, function(x) sum(x == 0) <= 70)
sum(keep) # 21,355
sum(!keep) # 15,357

counts.tpm <- counts.tpm[keep,]

# Check how this filter affects cases and controls:
cases.tpm <- counts.tpm[,1:97]
controls.tpm <- counts.tpm[,98:ncol(counts.tpm)]

#sum(apply(cases.tpm, 1, function(x) sum(x == 0) <= 48)) # 21,197
#sum(apply(controls.tpm, 1, function(x) sum(x == 0) <= 22)) # 20,499
```

### Differential Expression by Size of the Aneurysm size:

```{r}
print(paste0("There are ", sum(is.na(bio$aortc_diameter_mm))," NAs"))
#table(bio$aortc_diameter_mm, useNA = "always")

Pvalue <- apply(dge$counts, 1, function(x) summary(lm(x~bio$aortc_diameter_mm+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,4])
Coef <- apply(dge$counts, 1, function(x) summary(lm(x~bio$aortc_diameter_mm+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,1])
```

```{r}
# Histogram of p-values:
hist(Pvalue)
```

### Create qvalue object:

```{r}
qobj <- qvalue(p = Pvalue)
```

```{r, warning=FALSE}
summary(qobj)
hist(qobj)
plot(qobj)
```

```{r}
Results.size <- data.frame(
  pvalue = Pvalue,
  qvalue = qobj$qvalues,
  fdr = p.adjust(Pvalue, method = "fdr"),
  pi1 = 1-qobj$pi0
)

Results.size <- merge(Results.size, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
Results.size <- Results.size[,-1]
Results.size <- Results.size[!duplicated(Results.size$gene_name),]
Results.size <- Results.size %>% remove_rownames %>% column_to_rownames(var="gene_name")

Results.size[order(Results.size$pvalue), ][1:30,]
```


### Differential Expression by Age:

```{r}
print(paste0("There are ", sum(is.na(bio$age))," NAs"))
#table(bio$age, useNA = "always")

Pvalue <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$age+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,4])
Coef <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$age+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,1])
```

```{r}
# Histogram of p-values:
hist(Pvalue)
```

### Create qvalue object:

```{r}
qobj <- qvalue(p = Pvalue)
```

```{r, warning=FALSE}
summary(qobj)
hist(qobj)
plot(qobj)
```

```{r}
Results.age <- data.frame(
  pvalue = Pvalue,
  qvalue = qobj$qvalues,
  fdr = p.adjust(Pvalue, method = "fdr"),
  pi1 = 1-qobj$pi0
)

Results.age <- merge(Results.age, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
Results.age <- Results.age[,-1]
Results.age <- Results.age[!duplicated(Results.age$gene_name),]
Results.age <- Results.age %>% remove_rownames %>% column_to_rownames(var="gene_name")

Results.age[order(Results.age$pvalue), ][1:20,]
```

### Differential Expression by Sex (1 Men 2 Women):

```{r}
print(paste0("There are ", sum(is.na(bio$SEXO))," NAs"))
table(bio$SEXO, useNA = "always")

Pvalue <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$SEXO+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,4])
Coef <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$SEXO+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,1])
```

```{r}
# Histogram of p-values:
hist(Pvalue)
```

### Create qvalue object:

```{r}
qobj <- qvalue(p = Pvalue)
```

```{r, warning=FALSE}
summary(qobj)
hist(qobj)
plot(qobj)
```

```{r}
Results.sex <- data.frame(
  pvalue = Pvalue,
  qvalue = qobj$qvalues,
  fdr = p.adjust(Pvalue, method = "fdr"),
  pi1 = 1-qobj$pi0
)

Results.sex <- merge(Results.sex, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
Results.sex <- Results.sex[,-1]
Results.sex <- Results.sex[!duplicated(Results.sex$gene_name),]
Results.sex <- Results.sex %>% remove_rownames %>% column_to_rownames(var="gene_name")

Results.sex[order(Results.sex$pvalue), ][1:20,]
```

### Differential Expression by Smoking status (0 Never 1 Past (+6 months) 2 Current): NOTHING.


### Differential Expression by Type of Aneurysm (1 Atherosclerotic 2 Inflammatory):

```{r}
print(paste0("There are ", sum(is.na(bio$TIPOANEURISMA))," NAs"))
table(bio$TIPOANEURISMA)

Pvalue <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$TIPOANEURISMA+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,4])
Coef <- apply(cases.tpm, 1, function(x) summary(lm(x~bio$TIPOANEURISMA+tec$Date+tec$GC_Mean+tec$Batch))[4][[1]][2,1])
```

```{r}
# Histogram of p-values:
hist(Pvalue)
```

### Create qvalue object:

```{r}
qobj <- qvalue(p = Pvalue)
```

```{r, warning=FALSE}
summary(qobj)
hist(qobj)
plot(qobj)
```

```{r}
Results.type <- data.frame(
  pvalue = Pvalue,
  qvalue = qobj$qvalues,
  fdr = p.adjust(Pvalue, method = "fdr"),
  pi1 = 1-qobj$pi0
)

Results.type <- merge(Results.type, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
Results.type <- Results.type[,-1]
Results.type <- Results.type[!duplicated(Results.type$gene_name),]
Results.type <- Results.type %>% remove_rownames %>% column_to_rownames(var="gene_name")

Results.type[order(Results.type$pvalue), ][1:20,]
```

## Limma Analysis:

### Filter lowely expressed genes:

Keep genes that have more than 50% of non-zero counts across all samples:
```{r}
keep <- apply(counts.tpm, 1, function(x) sum(x == 0) <= 70)
sum(keep) # 21,355

counts.tpm <- counts.tpm[keep,]

# Check how this filter affects cases and controls:
cases.tpm <- counts.tpm[,1:97]
controls.tpm <- counts.tpm[,98:ncol(counts.tpm)]

#sum(apply(cases.tpm, 1, function(x) sum(x == 0) <= 48)) # 21,197
#sum(apply(controls.tpm, 1, function(x) sum(x == 0) <= 22)) # 20,499
```

### DEG by Aneurysm Size:

```{r}
dge <- DGEList(counts=cases.tpm)

design <- model.matrix(~bio$aortc_diameter_mm+tec$Date+tec$GC_Mean+tec$Batch)

colnames(design) <- c("(Intercept)", "Size", "Date202220829", "GC_Mean", "Plate")
```

### Remove lowely expressed genes:

```{r}
keep <- filterByExpr(dge, design)

paste0("We keep ", sum(keep), " genes")
paste0("We remove ", sum(!keep), " genes")

dge <- dge[keep,,keep.lib.sizes=FALSE]
```

### Remove NAs:

```{r}
keep <- !is.na(bio$aortc_diameter_mm)

paste("There are", sum(!keep), "NAs")

dge2 <- dge[,keep]
```

### Apply TMM Normalization:

```{r}
dge2 <- calcNormFactors(dge2, method = "TMM")
```

```{r}
v <- voom(dge2, design, plot = TRUE)

# Limma Differential Expression:

fit <- lmFit(v, design)
fit <- eBayes(fit)

# Obtain differentially expressed genes between all conditions:

summary(decideTests(fit))
```

### DEG by Age:

```{r}
design <- model.matrix(~bio$age+tec$Date+tec$GC_Mean+tec$Batch)

colnames(design) <- c("(Intercept)", "Age", "Date202220829", "GC_Mean", "Plate")

v <- voom(dge, design, plot = TRUE)

# Limma Differential Expression:

fit <- lmFit(v, design)
fit <- eBayes(fit)

# Obtain differentially expressed genes between all conditions:

summary(decideTests(fit))
```

### DEG by Sex:

```{r}
design <- model.matrix(~bio$SEXO+tec$Date+tec$GC_Mean+tec$Batch)

colnames(design) <- c("(Intercept)", "Sex", "Date202220829", "GC_Mean", "Plate")

v <- voom(dge, design, plot = TRUE)

# Limma Differential Expression:

fit <- lmFit(v, design)
fit <- eBayes(fit)

# Obtain differentially expressed genes between all conditions:

summary(decideTests(fit))
```

### DEG by Smoking Status:

```{r}
keep <- !is.na(bio$Smoking)

dge2 <- dge[,keep]

design <- model.matrix(~bio$Smoking+tec$Date+tec$GC_Mean+tec$Batch)

colnames(design) <- c("(Intercept)","Smoking1","Smoking2","Date202220829", "GC_Mean", "Plate")

v <- voom(dge2, design, plot = TRUE)

# Limma Differential Expression:

fit <- lmFit(v, design)
fit <- eBayes(fit)

# Obtain differentially expressed genes between all conditions:

summary(decideTests(fit))
```

### DEG by Disease:

```{r}
table(bio$BMI.di, useNA = "always")
keep <- !is.na(bio$BMI.di)

dge2 <- dge[,keep]

design <- model.matrix(~bio$BMI.di+tec$Date+tec$GC_Mean+tec$Batch)

colnames(design) <- c("(Intercept)","Disease","Date202220829", "GC_Mean", "Plate")

v <- voom(dge2, design, plot = TRUE)

# Limma Differential Expression:

fit <- lmFit(v, design)
fit <- eBayes(fit)

# Obtain differentially expressed genes between all conditions:
summary(decideTests(fit))
```




















