---
title: "Limma Analysis Salmon Cases"
author: "Gerard Temprano Sagrera"
date: "2022-12-16"
output: html_document
---

```{r, warning=FALSE, message=FALSE, echo = FALSE}
library(data.table);library(edgeR);library(tidyverse);library(openxlsx);library(preprocessCore)
```

### Apply TPM transformation to salmon counts:

```{r}
# Load previously prepared salmon read counts:
counts <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.salmon.txt")
lengths <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Lengths.txt")
lengths <- lengths$length/1000
rnames <- counts$V1
counts <- counts[,-1]
rownames(counts) <- rnames

r_tpm <- function(dfr,len)
{
  dfr1 <- sweep(dfr,1,len,`/`)
  sc.f <- colSums(dfr1)/(10^6)
  return(sweep(dfr1,2,sc.f,`/`))
}

counts.tpm <- as.data.frame(r_tpm(counts, lengths))
```

### Select only cases on technical variables:

```{r}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_CompleteV2.xlsx")

# Select only cases:
tec <- tec[1:97,]
```

### Filter lowely expressed genes:

Keep genes that have more than 50% of non-zero counts across all samples:
```{r}
keep <- apply(counts.tpm, 1, function(x) sum(x == 0) <= 70)
#sum(keep) # 21,355

counts.tpm <- counts.tpm[keep,]

# Check how this filter affects cases and controls:
cases.tpm <- counts.tpm[,1:97]
controls.tpm <- counts.tpm[,98:ncol(counts.tpm)]

#sum(apply(cases.tpm, 1, function(x) sum(x == 0) <= 48)) # 21,197
#sum(apply(controls.tpm, 1, function(x) sum(x == 0) <= 22)) # 20,499
```

### Convert TPM counts to DGElist object:

```{r}
dge <- DGEList(counts=cases.tpm)
```

### Apply TMM Normalization:

```{r}
dge <- calcNormFactors(dge, method = "TMM")
```

### Load Biological data to make comparissons:

```{r}
bio <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/20221115_Datos Pacientes.xlsx")
bio <- bio[2:nrow(bio),]

bio <- bio[bio$Status == "Case",]

# Create BMI:
bio$BMI <- bio$Weight_Kg/((bio$Height_cm/100)^2)
bio$BMI.di <- ifelse(bio$BMI > 25, 1, 0)
bio$BMI.tri <- ifelse(bio$BMI > 30, 2, ifelse(bio$BMI > 25, 1, 0))


# Transform aortic diameter to dicotomic variable:
bio$aortc_diameter_mm2 <- ifelse(bio$aortc_diameter_mm > 55, 1, 0) 
```

### Create design matrix:

```{r}
design.sex <- model.matrix(~bio$SEXO+tec$Date+tec$GC_Mean+tec$Batch)

design.type <- model.matrix(~bio$TIPOANEURISMA+tec$Date+tec$GC_Mean+tec$Batch)

design.sympt <- model.matrix(~bio$symptomatic+tec$Date+tec$GC_Mean+tec$Batch)

design.bmi.di <- model.matrix(~bio$BMI.di+tec$Date+tec$GC_Mean+tec$Batch)

design.bmi.tri <- model.matrix(~bio$BMI.tri+tec$Date+tec$GC_Mean+tec$Batch)

design.smoke <- model.matrix(~bio$Smoking+tec$Date+tec$GC_Mean+tec$Batch)

design.diameter <- model.matrix(~bio$aortc_diameter_mm2+tec$Date+tec$GC_Mean+tec$Batch)

design.other.ann <- model.matrix(~bio$other_aneurysm+tec$Date+tec$GC_Mean+tec$Batch)
```

### Apply voom transformation:
THERE ARE NAS and does not work.
```{r}
designs <- c("design.sex","design.type","design.sympt","design.bmi.di","design.bmi.tri","design.smoke",
             "design.diameter","design.other.ann")
v.list <- c()

for (i in 1:length(designs)){}
v.list[[i]] <- voom(dge, get(designs[i]), plot = TRUE)
i=2
```

### Limma Differential Expression:

```{r}
fit <- lmFit(v, design)
fit <- eBayes(fit)
```

```{r}
lm <- topTable(fit, coef=1, number = Inf)

summary(decideTests(fit))
```

### Obtain Gene_ID from Ensembl_ID, using annotation file:

```{r}
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v34.annotation.fixed.gtf.gz")
lm.deg2 <- merge(lm.deg, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")

if(nrow(lm.deg) != nrow(lm.deg2)){print("Some genes have been lost")}
```























