---
title: "Manual Analysis Salmon"
author: "Gerard Temprano Sagrera"
date: "2022-12-12"
output: html_document
---

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(data.table);library(edgeR);library(tidyverse);library(openxlsx);library(preprocessCore);library("org.Hs.eg.db");library(stats);library(qvalue)
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
# Slow loading.
library(clusterProfiler)
```

```{r}
#browseVignettes(package = "qvalue")
```

### Load Annotation file to change format names:

```{r}
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v34.annotation.fixed.gtf.gz")
```

### Apply TPM transformation to salmon counts:

```{r}
# Load previously prepared salmon read counts:
counts <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.salmon.txt")
lengths <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Lengths.txt")
lengths <- lengths$length/1000
rnames <- counts$V1
counts <- counts[,-1]
rownames(counts) <- rnames

r_tpm <- function(dfr,len)
{
  dfr1 <- sweep(dfr,1,len,`/`)
  sc.f <- colSums(dfr1)/(10^6)
  return(sweep(dfr1,2,sc.f,`/`))
}

counts.tpm <- as.data.frame(r_tpm(counts, lengths))
```

### Create a case/control variable:

```{r}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_Complete.xlsx")
cc <- tec[,c("Case")]
cc[cc == 1] <- "Case"
cc[cc == 0] <- "Control"
```


### Load Biological data to make comparissons:

```{r}
bio <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/20221115_Datos Pacientes.xlsx")
bio <- bio[2:nrow(bio),]

# Smoking:
bio$Smoking2 <- ifelse(bio$Smoking == 0, 0, 1)
```

### Convert TPM counts to DGElist object:

```{r}
dge <- DGEList(counts = counts.tpm)

# Set "Control" to be the first level:
AAA <- factor(x = c(rep("Case", 97), rep("Control", 44)),
              levels = c("Control", "Case"))

# Batch and FC are not included because are highly correlated with AAA:

design <-
  model.matrix(~ AAA + tec$Lane + tec$GC_Mean + tec$RIN + tec$FC + tec$Library + bio$age)

colnames(design) <-
  c("(Intercept)",
    "AAACase",
    "Date202220829",
    "GC_Mean",
    "RIN",
    "FC",
    "Plate",
    "Age")
```

### Remove lowely expressed genes using LIMMA:

```{r}
keep <- filterByExpr(dge, design)

paste0("We keep ", sum(keep), " genes") # 10,342
paste0("We remove ", sum(!keep), " genes") # 26,370

dge <- dge[keep,,keep.lib.sizes=FALSE]
```

### Apply quantile normalization to counts:

```{r}
tpm.rows <- rownames(counts.tpm)
tpm.cols <- colnames(counts.tpm)

counts.tpm <- as.data.frame(normalize.quantiles(as.matrix(counts.tpm)))

rownames(counts.tpm) <- tpm.rows
colnames(counts.tpm) <- tpm.cols
```

### Manual Differential Expression:

```{r}
Pvalue <-
  apply(dge$counts, 1, function(x)
    summary(
      lm(
        x ~ tec$Case + tec$Date + tec$GC_Mean + tec$RIN + tec$FC + tec$Batch + bio$age
      )
    )[4][[1]][2, 4])
Coef <-
  apply(dge$counts, 1, function(x)
    summary(
      lm(
        x ~ tec$Case + tec$Date + tec$GC_Mean + tec$RIN + tec$FC + tec$Batch + bio$age
      )
    )[4][[1]][2, 1])
```

### Plot P-values:

```{r}
png(file = "C://Users/Gerard/Desktop/AAA/RNAseq/GitHub/Hist_Cases_vs_Controls.png", height = 10000, width = 10000, res = 1200)
hist(Pvalue, main = "", cex.main = 2)
dev.off()

hist(Pvalue, main = "Cases vs Controls", cex.main = 2)
```

### Create qvalue object:

```{r}
qobj <- qvalue(p = Pvalue)
```

```{r, warning=FALSE}
summary(qobj)
hist(qobj)
plot(qobj)
```

```{r}
Results <- data.frame(
  pvalue = Pvalue,
  qvalue = qobj$qvalues,
  fdr = p.adjust(Pvalue, method = "fdr"),
  pi1 = 1-qobj$pi0,
  Beta = Coef
)

Results <- merge(Results, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
Results <- Results[,-1]
Results <- Results[!duplicated(Results$gene_name),]
Results <- Results %>% remove_rownames %>% column_to_rownames(var="gene_name")

Results <- Results[order(Results$pvalue),]
#write.xlsx(Results[order(Results$pvalue), ][1:10,],"C://Users/Gerard/Desktop/AAA/RNAseq/TopTable.xlsx", rowNames = TRUE )

nrow(Results[Results$fdr < 0.05,])
Results[1:50,]
Results[Results$Beta > 0,]
```

### Plot residuals:

#### TBXA2R:

```{r}
counts <- merge(dge$counts, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
counts <- counts[,-1]
counts <- counts[!duplicated(counts$gene_name),]
counts <- counts %>% remove_rownames %>% column_to_rownames(var="gene_name")

counts <- as.matrix(counts)

res <- resid(lm(counts["TBXA2R",] ~ tec$Date + tec$GC_Mean + tec$RIN + tec$FC + tec$Batch + bio$age))
res <-  data.frame(Expression = res)

p <- ggplot(res, aes(x = AAA, y = Expression, fill = AAA)) + geom_violin() + ggtitle("TBXA2R") +
  theme(
    plot.title = element_text(hjust = 0.5, size = 22),
    legend.text = element_text(size = 18),
    legend.title = element_text(size = 20),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    axis.text.x = element_text(size = 20),
    axis.text.y = element_text(size = 20)
  ) +
  scale_fill_discrete(name = "Status")

p + scale_fill_manual(values=c("deepskyblue1", "tomato1"))
```

#### TPM3:

```{r}
counts <-
  merge(dge$counts, ann[, c("gene_id", "gene_name")], by.x = 0, by.y = "gene_id")
counts <- counts[, -1]
counts <- counts[!duplicated(counts$gene_name), ]
counts <-
  counts %>% remove_rownames %>% column_to_rownames(var = "gene_name")

counts <- as.matrix(counts)

res <-
  resid(lm(
    counts["TPM3", ] ~ tec$Date + tec$GC_Mean + tec$RIN + tec$FC + tec$Batch + bio$age
  ))
res <-  data.frame(Expression = res)

p <- ggplot(res, aes(x = AAA, y = Expression, fill = AAA)) + geom_violin() + ggtitle("TPM3") +
  theme(
    plot.title = element_text(hjust = 0.5, size = 22),
    legend.text = element_text(size = 18),
    legend.title = element_text(size = 20),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    axis.text.x = element_text(size = 20),
    axis.text.y = element_text(size = 20)
  ) + scale_fill_discrete(name = "Status")

p + scale_fill_manual(values = c("deepskyblue1", "tomato1"))
```

### Compare with previous results:

```{r}
Sig.Res <- Results[Results$fdr <= 0.05,]

Media <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Media.xlsx", colNames = FALSE)
Adv <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Adventitia.xlsx", colNames = FALSE)

length(intersect(Media$X2, rownames(Sig.Res))) # 2,489

length(intersect(Adv$X2, rownames(Sig.Res))) # 901
```


### Prepare list for DAVID:

```{r}
#write.table(Sig.Res2$Row.names, "C://Users/Gerard/Desktop/AAA/RNAseq/DAVID.List.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
```


```{r}
#Gene_list <- Sig.Res2[,2]
#names(Gene_list) <- rownames(Sig.Res2)
#Gene_list <- Gene_list[order(-Gene_list)]

#gse <- gseGO(Gene_list,
#             ont = "BP",
#             keyType = "ENSEMBL",
#             OrgDb = "org.Hs.eg.db",
#             eps = 1e-300)
```

```{r}
#as.data.frame(gse)
```


```{r}
#fit <- gseaplot(gse, geneSetID = 1)
```












