---
title: "Manual Analysis Salmon"
author: "Gerard Temprano Sagrera"
date: "2022-12-12"
output: html_document
---

```{r, warning=FALSE, message=FALSE, echo = FALSE}
library(data.table);library(edgeR);library(tidyverse);library(openxlsx);library(preprocessCore);library(clusterProfiler);library("org.Hs.eg.db")

library(stats)
```

### Apply TPM transformation to salmon counts:

```{r}
# Load previously prepared salmon read counts:
counts <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.salmon.txt")
lengths <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Lengths.txt")
lengths <- lengths$length/1000
rnames <- counts$V1
counts <- counts[,-1]
rownames(counts) <- rnames

r_tpm <- function(dfr,len)
{
  dfr1 <- sweep(dfr,1,len,`/`)
  sc.f <- colSums(dfr1)/(10^6)
  return(sweep(dfr1,2,sc.f,`/`))
}

counts.tpm <- as.data.frame(r_tpm(counts, lengths))
```

### Create a case/control variable:

```{r}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_CompleteV2.xlsx")
cc <- tec[,c("Case")]
cc[cc == 1] <- "Case"
cc[cc == 0] <- "Control"
```

### Apply quantile normalization to counts:

```{r}
tpm.rows <- rownames(counts.tpm)
tpm.cols <- colnames(counts.tpm)

counts.tpm <- as.data.frame(normalize.quantiles(as.matrix(counts.tpm)))

rownames(counts.tpm) <- tpm.rows
colnames(counts.tpm) <- tpm.cols
```

### Filter lowely expressed genes:

Keep genes that have more than 50% of non-zero counts across all samples:
```{r}
keep <- apply(counts.tpm, 1, function(x) sum(x == 0) <= 70)
#sum(keep) # 21,355

counts.tpm <- counts.tpm[keep,]

# Check how this filter affects cases and controls:
cases.tpm <- counts.tpm[,1:97]
controls.tpm <- counts.tpm[,98:ncol(counts.tpm)]

#sum(apply(cases.tpm, 1, function(x) sum(x == 0) <= 48)) # 21,197
#sum(apply(controls.tpm, 1, function(x) sum(x == 0) <= 22)) # 20,499
```

### Manual Differential Expression:

```{r}
Pvalue <- apply(counts.tpm, 1, function(x) summary(lm(x~tec$Case+tec$Date+tec$GC_Mean+tec$RIN))[4][[1]][2,4])
Coef <- apply(counts.tpm, 1, function(x) summary(lm(x~tec$Case+tec$Date+tec$GC_Mean+tec$RIN))[4][[1]][2,1])

#summary(lm(dge$counts[1,]~tec$Case+tec$Date+tec$GC_Mean+tec$RIN))[4][[1]][2,4]
```

```{r}
Results <- as.data.frame(cbind(Pvalue, Coef))

Sig.Res <- Results[p.adjust(Pvalue, method = "bonferroni") < 0.05,]
```

### Obtain Gene_ID from Ensembl_ID, using annotation file:

```{r}
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v34.annotation.fixed.gtf.gz")
Sig.Res2 <- merge(Sig.Res, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
Sig.Res2 <- Sig.Res2 %>% remove_rownames %>% column_to_rownames(var="Row.names")


if(nrow(Sig.Res) != nrow(Sig.Res2)){print("Some genes have been lost")}
```

### Compare with previous results:

```{r}
Media <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Media.xlsx", colNames = FALSE)
Adv <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Adventitia.xlsx", colNames = FALSE)

length(intersect(Media$X2, Sig.Res2$gene_name)) # 3,865

length(intersect(Adv$X2, Sig.Res2$gene_name)) # 1,423
```

### Prepare list for DAVID:

```{r}
write.table(Sig.Res2$Row.names, "C://Users/Gerard/Desktop/AAA/RNAseq/DAVID.List.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
```


```{r}
Gene_list <- Sig.Res2[,2]
names(Gene_list) <- rownames(Sig.Res2)
Gene_list <- Gene_list[order(-Gene_list)]

gse <- gseGO(Gene_list,
             ont = "BP",
             keyType = "ENSEMBL",
             OrgDb = "org.Hs.eg.db",
             eps = 1e-300)
```

```{r}
as.data.frame(gse)
```


```{r}
fit <- gseaplot(gse, geneSetID = 1)



```












