---
title: "RNAseq AAA Prepare Salmon Counts"
author: "Gerard Temprano Sagrera"
date: "2022-12-08"
output: html_document
---

```{r}
# Load annotation file:
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v34.annotation.fixed.gtf.gz")
# Load salmon counts:
merged <- readRDS("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/salmon.merged.gene_counts.rds")

# Quantified genes:
ff <- rownames(merged)

# Select quantified genes:
genes2 <- ann[ann$gene_id%in%ff,]
# Select by gene type:
genes2 <- genes2[genes2$gene_type == "proteincoding" | genes2$gene_type == "lncRNA",]
# Select by annotation level:
genes2 <- genes2[genes2$level == 1 | genes2$level == 2,]
# Remove genes with the same ID:
genes2 <- genes2[!(duplicated(genes2$gene_id) | duplicated(genes2$gene_id, fromLast = TRUE)), ]
# Order by gene id:
genes2 <- genes2[order(genes2$gene_id),]
```

Save salmon counts and length of the genes:

```{r}
counts <- assays(merged)$counts
# Select by annotation and type:
counts <- counts[rownames(counts)%in%genes2$gene_id,]
colnames(counts) <- gsub("_R1","",colnames(counts))
counts <- counts[order(rownames(counts)),]
dim(counts)

identical(genes2$gene_id,rownames(counts))

fwrite(genes2[,c("gene_id","length")], "C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Lengths.txt", sep = "\t")
fwrite(counts, "C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.salmon.txt", sep = "\t", row.names = T)
```



