---
title: "RNAseq AAA Prepare Salmon Counts"
author: "Gerard Temprano Sagrera"
date: "2022-12-08"
output: html_document
---

```{r}
library(data.table)
```

```{r}
# Load annotation file:
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v34.annotation.fixed.gtf.gz")
ann$gene_id <- substr(ann$gene_id, 1, 15)

# Load salmon counts:
merged <- readRDS("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/salmon.merged.gene_counts.rds")
```

### Save salmon counts and length of the genes:

# Use abundance instead of counts, to obtain directly TPM.
```{r}
counts <- assays(merged)$abundance
dim(counts) # 60,666 141

# Select by annotation and type:
counts <- counts[rownames(counts) %in% ann$gene_id,]
colnames(counts) <- gsub("_R1","",colnames(counts))
counts <- counts[order(rownames(counts)),]

dim(counts) # 36,742 141

#fwrite(counts, "C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.Salmon.V34.txt", sep = "\t", row.names = T)
```



