---
title: "1.3.Manual_Analysis_RSEM"
author: "Gerard Temprano Sagrera"
date: "2023-02-24"
output: html_document
---

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(data.table);library(edgeR);library(tidyverse);library(openxlsx);library(preprocessCore);library("org.Hs.eg.db");library(stats);library(qvalue);library(ggpubr);library(VennDiagram)
```

### Load Annotation file to change format names:

```{r}
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v26.annotation.fixed.gtf.gz")
```

### Counts are already in TPM from RSEM:

```{r}
counts.tpm <- read.table("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.RSEM.All.txt")
```

```{r}
tpm_matrix <- as.matrix(counts.tpm)
#top10_genes_tpm <- tpm_matrix[1:10, ]

#for (i in 1:nrow(top10_genes_tpm)) {
#  gene_tpm <- top10_genes_tpm[i, ]
#  gene_name <- rownames(top10_genes_tpm)[i]
#  plot_data <- data.frame(Sample = colnames(top10_genes_tpm), TPM = gene_tpm)
#  
#  p <- ggplot(plot_data, aes(x = TPM)) +
#    geom_density() +
#    ggtitle(paste("Density plot for gene", gene_name))
#  
#  print(p)
#}
```

### Load technical table :

```{r}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_Complete.xlsx")
tec <- tec[tec$Seq_ID %in% colnames(counts.tpm),]
```

### Load Biological data:

```{r}
bio <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/20221115_Datos Pacientes.xlsx")
bio <- bio[2:nrow(bio),]
bio <- bio[bio$Muestra %in% colnames(counts.tpm),]

table(bio$Status)

# Smoking:
bio$Smoking2 <- ifelse(bio$Smoking == 0, 0, 1)
```


### Convert TPM counts to DGElist object:

```{r}
dge <- DGEList(counts = counts.tpm)

# Set "Control" to be the first level:
AAA <- factor(x = c(rep("Case", 96), rep("Control", 44)),
              levels = c("Control", "Case"))

# Batch and FC are not included because are highly correlated with AAA:
design <-
  model.matrix(~ AAA + tec$FC + tec$Lane + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$SEXO + bio$age)

colnames(design) <-
  c("(Intercept)",
    "AAA",
    "FC1",
    "FC2",
    "Lane1",
    "Lane2",
    "GC_Mean",
    "RIN",
    "DV200",
    "Qubit",
    "Sex",
    "Age")
```

### Remove lowely expressed genes using manual filtering:

```{r}
keep <- apply(dge$counts, 1, function(x) sum(x > 0.5) >= 70)
dge <- dge[keep,,keep.lib.sizes=FALSE]

paste0("We keep ", sum(keep), " genes") # 14,675
paste0("We remove ", sum(!keep), " genes") # 12,615 
```

### Apply quantile normalization to counts:

```{r}
counts.tpm <- as.data.frame(normalize.quantiles(as.matrix(dge$counts), keep.names = T))
```


```{r}
tpm_matrix <- as.matrix(counts.tpm)

#top10_genes_tpm <- tpm_matrix[1:10, ]

#for (i in 1:nrow(top10_genes_tpm)) {
#  gene_tpm <- top10_genes_tpm[i, ]
#  gene_name <- rownames(top10_genes_tpm)[i]
#  plot_data <- data.frame(Sample = colnames(top10_genes_tpm), TPM = gene_tpm)
#  
#  p <- ggplot(plot_data, aes(x = TPM)) +
#    geom_density() +
#    ggtitle(paste("Density plot for gene", gene_name))
  
#  print(p)
#}
```


### Manual Differential Expression:

```{r}
Pvalue <-
  apply(counts.tpm, 1, function(x)
    summary(
      lm(
        x ~ tec$Case + tec$FC + tec$Lane + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$age + bio$SEXO
      )
    )[4][[1]][2, 4])

Coef <-
  apply(counts.tpm, 1, function(x)
    summary(
      lm(
        x ~ tec$Case + tec$FC + tec$Lane + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$age + bio$SEXO
      )
    )[4][[1]][2, 1])
```


### Plot P-values:

```{r}
png(file = "C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Results/Hist_Cases_vs_Controls.png", height = 10000, width = 10000, res = 1200)
hist(Pvalue, main = "Cases vs Controls", cex.main = 2)
dev.off()

hist(Pvalue, main = "Cases vs Controls", cex.main = 2)
```

### Create qvalue object:

```{r}
qobj <- qvalue(p = Pvalue)
```

```{r, warning=FALSE}
#summary(qobj)
#hist(qobj)
#plot(qobj)
```

```{r}
Results <- data.frame(
  pvalue = Pvalue,
  qvalue = qobj$qvalues,
  fdr = p.adjust(Pvalue, method = "fdr"),
  pi1 = 1-qobj$pi0,
  Beta = Coef
)

Results <- merge(Results, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
names(Results)[names(Results) == "Row.names"] <- "gene_id"
Results <- Results[!duplicated(Results$gene_name),]
Results <- Results %>% remove_rownames %>% column_to_rownames(var="gene_name")
Results <- Results[order(Results$pvalue),]

nrow(Results[Results$fdr < 0.05,]) # 7,454

head(Results)

glist <- c("PDP1","PDP2","PDHA1","PDK1","PDK2","PDK3","PDK4","ACO1","ACO2","IDH1","IDH2","IDH3A","IDH3B","OGDH","SUCLA2",
           "SDHA","SDHB","SDHC","SDHD","GLS","GLUL","GLUD1","GAD1","GAD2","ABAT","ALDH5A1")
           
#write.xlsx(Results[rownames(Results) %in% glist,], "C://Users/Gerard/Desktop/AAA/RNAseq/Genes.xlsx", rowNames = T)

#write.table(rownames(head(Results, 50)), "C://Users/Gerard/Desktop/Top50Genes.txt", quote = F, rowNames = F, col.names = F)

#write.xlsx(Results[Results$fdr < 0.05,], "C://Users/Gerard/Desktop/AAA/RNAseq/SigRes/Cases_Controls_No_Smoking.xlsx", sep = "\t", rowNames = T)
```

### Compare with previous results (Tunica-Specific):

```{r}
# Load our results:
cc <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/SigRes/Cases_Controls_No_Smoking.xlsx")
```

```{r}
med <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/TunicaSpecificResults.xlsx", sheet = 1)
med$Gene.Symbol <- gsub("-", "", med$Gene.Symbol)

#length(med$Gene.Symbol) # 5,889
#length(unique(med$Gene.Symbol)) # 5,507

#length(intersect(med$Gene.Symbol, cc$V1)) # 2,952

#length(setdiff(med$Gene.Symbol, cc$V1)) # 2,555
#length(setdiff(cc$V1, med$Gene.Symbol)) # 2,840
```

```{r}
adv <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/TunicaSpecificResults.xlsx", sheet = 2)
adv$Gene.Symbol <- gsub("-", "", adv$Gene.Symbol)

#length(adv$Gene.Symbol) # 2,701
#length(unique(adv$Gene.Symbol)) #2,508

#length(intersect(adv$Gene.Symbol, cc$V1)) # 1,007

#length(setdiff(adv$Gene.Symbol, cc$V1)) # 1,501
#length(setdiff(cc$V1, adv$Gene.Symbol)) # 4,785
```

```{r}
med <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/TunicaSpecificResults.xlsx", sheet = 1)
med$Gene.Symbol <- gsub("-", "", med$Gene.Symbol)

adv <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/TunicaSpecificResults.xlsx", sheet = 2)
adv$Gene.Symbol <- gsub("-", "", adv$Gene.Symbol)

updown <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/UpDown Results.xlsx")

smla <- readxl::read_xls("C://Users/Gerard/Desktop/AAA/RNAseq/SmallLargeResults.xls")

rup <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Ruptured.xlsx")

affyilu <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Affyilu_Results.xlsx")


genes <- fread("C://Users/Gerard/Desktop/Top50Genes.txt", header = F)


# Unifying gene lists
df_list <- list(adv = list(df = adv, code = "32907367"),
                med = list(df = med, code = "32907367"),
                updown = list(df = updown, code = "19111481"),
                smla = list(df = smla, code = "25944698"),
                rup = list(df = rup, code = "29191809"),
                affyilu = list(df = affyilu, code = "17634102"))

# Create new dataframe:
gene_presence <- data.frame(Gene.Name = genes$V1, Symbol_Code = as.character(rep("", length(genes$V1))))

# Check gene presence:
for (df_name in names(df_list)) {
  df_info <- df_list[[df_name]]
  in_gene_index <- gene_presence$Gene.Name %in% df_info$df$Gene.Symbol
  gene_presence[in_gene_index, "Symbol_Code"] <- ifelse(gene_presence[in_gene_index, "Symbol_Code"] == "", 
                                                        df_info$code,
                                                        paste(gene_presence[in_gene_index, "Symbol_Code"], df_info$code, sep = ","))
}

names(gene_presence) <- c("Gene Name","Previously Described")
gene_presence$'Functional Studies' <- NA

#write.xlsx(gene_presence, "C://Users/Gerard/Desktop/AAA/RNAseq/Top50Genes.xlsx")
```


### DEG:

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(clusterProfiler);library(org.Hs.eg.db);library(enrichplot);library(ggplot2)
```

```{r}
gene_list <- substr(Results[Results$fdr < 0.05,]$gene_id, 1, 15)

entrez_ids <- bitr(gene_list, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db)

# GO enrichment analysis:
go_enrich <- enrichGO(gene          = entrez_ids$ENTREZID,
                      OrgDb         = org.Hs.eg.db,
                      keyType       = "ENTREZID",
                      ont           = "ALL", # Choose from "BP" (Biological Process), "CC" (Cellular Component), or "MF" (Molecular Function)
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff  = 0.2)
```

```{r}
# GO enrichment dot plot:
#png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Enrichment/GO_Cases_vs_Controls_No_Smoking.png", width = 4*800, height = 4*600, res = 300)
dotplot(go_enrich, showCategory = 20) + ggtitle("GO Enrichment")
#dev.off()
```

```{r}
gene_list <- rownames(Results[Results$fdr < 0.05,])

# KEGG enrichment analysis
kegg_enrich <- enrichKEGG(gene         = entrez_ids$ENTREZID,
                          organism     = "hsa",
                          keyType      = "kegg",
                          pAdjustMethod = "BH",
                          pvalueCutoff = 0.05,
                          qvalueCutoff = 0.2)
```

```{r}
# KEGG enrichment dot plot
#png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Enrichment/KEGG_Cases_vs_Controls_No_Smoking.png", width = 4*800, height = 4*600, res = 300)
dotplot(kegg_enrich, showCategory = 15) + ggtitle("KEGG Enrichment")
#dev.off()
```


# STRING and Cytoscape integration:

```{r}
library(RCy3)
cytoscapePing()
cytoscapeVersionInfo()

installApp('stringApp')

de.genes <- read.table("https://raw.githubusercontent.com/cytoscape/cytoscape-tutorials/gh-pages/protocols/data/TCGA-Ovarian-MesenvsImmuno_data.csv", header = TRUE, sep = ",", quote="\"", stringsAsFactors = FALSE)

de.genes.up <- de.genes[which(de.genes$logFC >= 1.8 & de.genes$FDR.adjusted.Pvalue <= 0.05),]

string.cmd = paste('string protein query query="', paste(de.genes.up$Gene, collapse = '\n'), '" cutoff=0.4 limit=0 species="Homo sapiens"', sep = "")

#commandsRun(string.cmd)
```




### Volcano Plot:

```{r}
significant_genes <- Results %>% filter(qvalue < 0.0005)

g<- ggplot(data = Results, aes(x = Beta, y = -log(pvalue))) +
  geom_point(aes(color = qvalue < 0.05), size = 0.5) +
  scale_color_manual(values = c("gray", "red")) +
  geom_hline(yintercept = -log10(0.05)) +
  geom_vline(xintercept = 1) +
  geom_vline(xintercept = -1) +
#  xlim(-20,20) +
  theme_minimal() +
  geom_text(data = significant_genes,
            aes(label = rownames(significant_genes), hjust = 0, vjust = -0.5),
            size = 3, check_overlap = TRUE)

png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Results/Volcano_Cases_vs_Controls_No_Smoking", width = 4*800, height = 4*600, res = 300)
g
dev.off()
```


### Plot residuals:

#### ERBB2:

##### By Case:

```{r}
counts <- merge(counts.tpm, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
counts <- counts[,-1]
counts <- counts[!duplicated(counts$gene_name),]
counts <- counts %>% remove_rownames %>% column_to_rownames(var="gene_name")

counts <- as.matrix(counts)
```

```{r}
cases.tpm <- counts[,1:96]
controls.tpm <- counts[,97:ncol(counts)]

dim(cases.tpm)
dim(controls.tpm)

mean(cases.tpm["BTC",])
mean(controls.tpm["BTC",])

table(AAA)
```

```{r}
res <- resid(lm(counts["BTC",] ~ tec$FC + tec$Lane + tec$Library + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit))
res <-  data.frame(Expression = res)


p <- ggplot(res, aes(x = AAA, y = Expression, fill = AAA)) + geom_boxplot() + ggtitle("BTC") +
  stat_compare_means(method = "t.test", hjust = -0.5) + ylab("Expression") +
  theme(
    plot.title = element_text(hjust = 0.5, size = 22),
    legend.text = element_text(size = 18),
    legend.title = element_text(size = 20),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    axis.text.x = element_text(size = 20),
    axis.text.y = element_text(size = 20)
  ) +
  scale_fill_discrete(name = "Status") +
  scale_fill_manual(values=c("cyan3", "red"))

p

#ggsave("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/ERBB4.png", width = 7, height = 5)
```



#### EGFR:

##### By Case:

```{r}
counts <- merge(counts.tpm, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
counts <- counts[,-1]
counts <- counts[!duplicated(counts$gene_name),]
counts <- counts %>% remove_rownames %>% column_to_rownames(var="gene_name")

counts <- as.matrix(counts)
```

```{r}
cases.tpm <- counts[,1:96]
controls.tpm <- counts[,97:ncol(counts)]

dim(cases.tpm)
dim(controls.tpm)

mean(cases.tpm["EGFR",])
mean(controls.tpm["EGFR",])

table(AAA)
```

```{r}
res <- resid(lm(counts["EGFR",] ~ tec$FC + tec$Lane + tec$Library + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit))
res <-  data.frame(Expression = res)


p <- ggplot(res, aes(x = AAA, y = Expression, fill = AAA)) + geom_boxplot() + ggtitle("EGFR") +
  stat_compare_means(method = "t.test", hjust = -0.5) + ylab("Expression (Normalized TPM)") +
  theme(
    plot.title = element_text(hjust = 0.5, size = 22),
    legend.text = element_text(size = 18),
    legend.title = element_text(size = 20),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    axis.text.x = element_text(size = 20),
    axis.text.y = element_text(size = 20)
  ) +
  scale_fill_discrete(name = "Status") +
  scale_fill_manual(values=c("cyan3", "red"))

p

ggsave("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/EGFR.png", width = 7, height = 5)
```


##### By Diameter:

```{r}
counts <- merge(counts.tpm, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
counts <- counts[,-1]
counts <- counts[!duplicated(counts$gene_name),]
counts <- counts %>% remove_rownames %>% column_to_rownames(var="gene_name")

counts <- as.matrix(counts)
```

```{r}
dia <- bio[!is.na(bio$aortc_diameter_mm),][,c("Muestra","aortc_diameter_mm")]

counts <- counts[,colnames(counts) %in% dia$Muestra]
tec2 <- tec[tec$Seq_ID %in% colnames(counts),]

dia <- dia %>% remove_rownames %>% column_to_rownames(var="Muestra")

res <- resid(lm(counts["EGFR",] ~ tec2$FC + tec2$Lane + tec2$Library + tec2$GC_Mean + tec2$RIN + tec2$DV200 + tec2$Qubit))


res <- data.frame(Expression = res,
                  Diameter = dia)
```

```{r}
ggplot(res, aes(x = aortc_diameter_mm, y = Expression)) + geom_point(color = "orangered2") + ggtitle("EGFR") + ylab ("Expression") + xlab("Aortic Diameter (mm)") +
  geom_smooth(method = 'lm', formula = y ~ x, color = "black") + stat_cor(method = "pearson") + 
  theme(
    plot.title = element_text(hjust = 0.5, size = 22),
    legend.text = element_text(size = 18),
    legend.title = element_text(size = 20),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    axis.text.x = element_text(size = 20),
    axis.text.y = element_text(size = 20),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank(),
    axis.line = element_line(colour = "black")
  )


#ggsave("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/EGFR_Diameter.png")
```

