---
title: "1.3.Manual_Analysis_RSEM"
author: "Gerard Temprano Sagrera"
date: "2023-02-24"
output: html_document
---

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(data.table);library(edgeR);library(tidyverse);library(openxlsx);library(preprocessCore);library("org.Hs.eg.db");library(stats);library(qvalue);library(ggpubr);library(VennDiagram)
```

### Load Annotation file to change format names:

```{r}
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v26.annotation.fixed.gtf.gz")
```

### Counts are already in TPM from RSEM:

```{r}
counts.tpm <- read.table("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.RSEM.All.txt")
```


### Load technical table :

```{r}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_Complete.xlsx")
tec <- tec[tec$Seq_ID %in% colnames(counts.tpm),]
```

### Load Biological data:

```{r}
bio <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/20221115_Datos Pacientes.xlsx")
bio <- bio[2:nrow(bio),]
bio <- bio[bio$Muestra %in% colnames(counts.tpm),]

table(bio$Status)

table(bio[bio$Status == "Case",]$SEXO)

table(bio[bio$Status == "Control",]$SEXO)

head(bio)
table(bio$TIPOANEURISMA, useNA="always")

# Smoking:
bio$Smoking2 <- ifelse(bio$Smoking == 0, 0, 1)

names(bio)
bio2 <- bio[,c("Status","SEXO","age","Smoking","CVE","med_antihypertensive","med_statins","med_anticoagulants")]

bio3 <- na.omit(bio2)

table(bio3$Status)
```


### Convert TPM counts to DGElist object:

```{r}
dge <- DGEList(counts = counts.tpm)

# Set "Control" to be the first level:
AAA <- factor(x = c(rep("Case", 96), rep("Control", 44)),
              levels = c("Control", "Case"))
```

### Remove lowely expressed genes using manual filtering:

```{r}
keep <- apply(dge$counts, 1, function(x) sum(x > 0.5) >= 70)
dge <- dge[keep,,keep.lib.sizes=FALSE]

paste0("We keep ", sum(keep), " genes") # 14,675
paste0("We remove ", sum(!keep), " genes") # 12,615
```


### Apply quantile normalization to counts:

```{r}
counts.tpm <- as.data.frame(normalize.quantiles(as.matrix(dge$counts), keep.names = T))

table(bio$Status, bio$med_statins)

```


```{r}
#tpm_matrix <- as.matrix(counts.tpm)

#top10_genes_tpm <- tpm_matrix[1:10, ]

#for (i in 1:nrow(top10_genes_tpm)) {
#  gene_tpm <- top10_genes_tpm[i, ]
#  gene_name <- rownames(top10_genes_tpm)[i]
#  plot_data <- data.frame(Sample = colnames(top10_genes_tpm), TPM = gene_tpm)
#  
#  p <- ggplot(plot_data, aes(x = TPM)) +
#    geom_density() +
#    ggtitle(paste("Density plot for gene", gene_name))
  
#  print(p)
#}
```


### Manual Differential Expression:

```{r}
Pvalue <-
  apply(counts.tpm, 1, function(x)
    summary(
      lm(
        x ~ tec$Case + tec$FC + tec$Lane + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$age + bio$SEXO
      )
    )[4][[1]][2, 4])

Coef <-
  apply(counts.tpm, 1, function(x)
    summary(
      lm(
        x ~ tec$Case + tec$FC + tec$Lane + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$age + bio$SEXO
      )
    )[4][[1]][2, 1])
```



### Manual Differential Expression with SIGNIFICANT CONFOUNDERS:

```{r}
Pvalue <-
  apply(counts.tpm, 1, function(x)
    summary(
      lm(
        x ~ tec$Case + tec$FC + tec$Lane + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$age + bio$SEXO + bio$Smoking + bio$CVE + bio$med_antihypertensive + bio$med_statins + bio$med_anticoagulants
      )
    )[4][[1]][2, 4])

Coef <-
  apply(counts.tpm, 1, function(x)
    summary(
      lm(
        x ~ tec$Case + tec$FC + tec$Lane + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$age + bio$SEXO + bio$Smoking + bio$CVE + bio$med_antihypertensive + bio$med_statins + bio$med_anticoagulants
      )
    )[4][[1]][2, 1])
```



### Plot P-values:

```{r}
#png(file = "C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Results/Hist_Cases_vs_Controls.png", height = 10000, width = 10000, res = 1200)
#hist(Pvalue, main = "", cex.main = 2)
#dev.off()

hist(Pvalue, main = "", cex.main = 2)
```

### Create qvalue object:

```{r}
qobj <- qvalue(p = Pvalue)
```

```{r, warning=FALSE}
#summary(qobj)
#hist(qobj)
#plot(qobj)
```

```{r}
Results <- data.frame(
  pvalue = Pvalue,
  qvalue = qobj$qvalues,
  fdr = p.adjust(Pvalue, method = "fdr"),
  pi1 = 1-qobj$pi0,
  Beta = Coef
)

Results <- merge(Results, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
names(Results)[names(Results) == "Row.names"] <- "gene_id"
Results <- Results[!duplicated(Results$gene_name),]
Results <- Results %>% remove_rownames %>% column_to_rownames(var="gene_name")
Results <- Results[order(Results$pvalue),]

nrow(Results[Results$fdr < 0.05,])

# 7,454 DEGs correcting for technical and AGE and SEX.
# 5,842 DEGs correcting for SIGNIFICANT CONFOUNDERS.

#write.xlsx(Results, "C://Users/Gerard/Desktop/AAA/RNAseq/AllRes/Cases_Controls.xlsx", quote = F, rowNames = T, colNames = T)

#write.xlsx(Results, "C://Users/Gerard/Desktop/AAA/RNAseq/AllRes/Cases_Controls_Confounders.xlsx", quote = F, rowNames = T, colNames = T)


#write.xlsx(Results[rownames(Results) %in% glist,], "C://Users/Gerard/Desktop/AAA/RNAseq/Genes.xlsx", rowNames = T)
#write.table(rownames(head(Results, 50)), "C://Users/Gerard/Desktop/Top50Genes.txt", quote = F, rowNames = F, col.names = F)

#write.xlsx(Results[Results$fdr < 0.05,], "C://Users/Gerard/Desktop/AAA/RNAseq/SigRes/Cases_Controls_No_Smoking.xlsx", sep = "\t", rowNames = T)

#write.xlsx(Results[Results$fdr < 0.05,], "C://Users/Gerard/Desktop/AAA/RNAseq/SigRes/Cases_Controls_Confounders.xlsx", sep = "\t", rowNames = T)
```



### DEG:

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(clusterProfiler);library(org.Hs.eg.db);library(enrichplot);library(ggplot2);library(enrichplot);library(ggupset);library(ggtree);library(aPEAR)
```


```{r}
gene_list <- substr(Results[Results$fdr < 0.05,]$gene_id, 1, 15)

entrez_ids <- bitr(gene_list, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db)

# GO enrichment analysis:
go_enrich <- enrichGO(gene          = entrez_ids$ENTREZID,
                      OrgDb         = org.Hs.eg.db,
                      keyType       = "ENTREZID",
                      ont           = "ALL", # Choose from "BP" (Biological Process), "CC" (Cellular Component), or "MF" (Molecular Function)
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff  = 0.2)

head(go_enrich, 3)
```



### Save results:

```{r}
#write.xlsx(go_enrich@result, "C://Users/Gerard/Desktop/AAA/RNAseq/Supplementary_Tables/GO_Enrich_Confounders.xlsx")

go_enrich@result <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Supplementary_Tables/GO_Enrich_Confounders.xlsx")
```



```{r}
clusters <- findPathClusters(go_enrich@result, cluster = 'hier', minClusterSize = 10)
unique(clusters$clusters$Cluster)
it.clusters <- unique(clusters$clusters$Cluster)
```


```{r}
set.seed(5)

pdf("C://Users/Gerard/Desktop/Clusters.pdf", height = 12, width = 12)
plot <- plotPathClusters(
  enrichment = go_enrich@result,
  sim = clusters$similarity,
  clusters = clusters$clusters,
  fontSize = 8,
  outerCutoff = 0.2,
  drawEllipses = TRUE,
  nodeSize = "Count",
  colorBy = "pvalue",
  colorType = 'pval',
  repelLabels = TRUE
)

legend(plot, title = NULL, cex = 1.5)  # Adjust cex value for desired size

dev.off()
```



```{r}
# GO enrichment dot plot:
#png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Enrichment/GO_Cases_vs_Controls_No_Smoking.png", width = 4*800, height = 4*600, res = 300)
dotplot(go_enrich, showCategory = 20) + ggtitle("GO Enrichment") +
  theme(plot.title = element_text(hjust = 0.5))
#dev.off()
```

```{r}
# KEGG enrichment analysis
kegg_enrich <- enrichKEGG(gene         = entrez_ids$ENTREZID,
                          organism     = "hsa",
                          keyType      = "kegg",
                          pAdjustMethod = "BH",
                          pvalueCutoff = 0.05,
                          qvalueCutoff = 0.2)
```


### Save results:

```{r}
#write.xlsx(kegg_enrich@result, "C://Users/Gerard/Desktop/AAA/RNAseq/Supplementary_Tables/KEGG_Enrich.xlsx")
```


```{r}
# KEGG enrichment dot plot
png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Enrichment/KEGG_Cases_vs_Controls_No_Smoking.png", width = 4*800, height = 4*600, res = 300)
dotplot(kegg_enrich, showCategory = 15) + ggtitle("KEGG Enrichment") +
  theme(plot.title = element_text(hjust = 0.5))
dev.off()
```


```{r}
edox <- setReadable(go_enrich, 'org.Hs.eg.db', 'ENTREZID')
p1 <- cnetplot(edox, foldChange=entrez_ids$ENTREZID)
p2 <- cnetplot(edox, categorySize="pvalue", foldChange=entrez_ids$ENTREZID)
p3 <- cnetplot(edox, foldChange=entrez_ids$ENTREZID, circular = TRUE, colorEdge = TRUE)
```


```{r}
png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Enrichment/CnetPlot2.png", height = 8000, width = 8000, res = 300)
p3
dev.off()
```


```{r, warning=F, echo=F, message=F}
edox2 <- pairwise_termsim(edox)
edox2
p2 <- treeplot(edox2, showCategory = 200, nCluster = 5)
```

```{r}
png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Enrichment/TreePlot_All2.png", height = 6000, width = 6000, res = 300)
p2
dev.off()
```

```{r}
edo <- pairwise_termsim(go_enrich)
p4 <- emapplot(edo, cex_category=1.5,layout="kk", showCategory = 200) 
```


```{r}
png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Enrichment/EnrichmentMap.png", height = 2000, width = 2000)
p4
dev.off()
```




### DEG No IT:

# Load ischemic time differentially expressed genes:

```{r}
isch <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Ischemic/Ichemic_Time_Genes.xlsx")
```


```{r}
# Remove ischemic time genes:
Results_No_IT <- Results[!rownames(Results) %in% isch$Gene_Name,]
nrow(Results_No_IT[Results_No_IT$fdr < 0.05,]) # 3,002
nrow(Results_No_IT)

# 2374 DEGs without Ischemic Time effect, correcting for SIGNIFICANT CONFOUNDERS.

#write.xlsx(Results, "C://Users/Gerard/Desktop/AAA/RNAseq/AllRes/Cases_Controls_NoIT.xlsx", quote = F, rowNames = T, colNames = T)

#write.xlsx(Results_No_IT[Results_No_IT$fdr < 0.05,], "C://Users/Gerard/Desktop/AAA/RNAseq/SigRes/Cases_Controls_Confounders_No_IT.xlsx", sep = "\t", rowNames = T)
```


### Plot P-values:

```{r}
#png(file = "C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Results/Hist_Cases_vs_Controls_NoIT.png", height = 10000, width = 10000, res = 1200)
#hist(Results_No_IT$pvalue, main = "", xlab = "Pvalue")
#dev.off()

hist(Results_No_IT$pvalue, main = "", xlab = "Pvalue")
```


```{r}
qobj <- qvalue(p = Results_No_IT$pvalue)

1-qobj$pi0
```



```{r}
gene_list <- substr(Results_No_IT[Results_No_IT$fdr < 0.05,]$gene_id, 1, 15)

entrez_ids <- bitr(gene_list, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db)

# GO enrichment analysis:
go_enrich_no_it <- enrichGO(gene    = entrez_ids$ENTREZID,
                      OrgDb         = org.Hs.eg.db,
                      keyType       = "ENTREZID",
                      ont           = "ALL", # Choose from "BP" (Biological Process), "CC" (Cellular Component), or "MF" (Molecular Function)
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff  = 0.2,
                      readable = T)
```


### Save results:

```{r}
#write.xlsx(go_enrich_no_it@result, "C://Users/Gerard/Desktop/AAA/RNAseq/Supplementary_Tables/GO_Enrich_NO_IT_Confounders.xlsx")

go_enrich_no_it@result <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Supplementary_Tables/GO_Enrich_NO_IT_Confounders.xlsx")
```



```{r}
clusters_no_it <- findPathClusters(go_enrich_no_it@result, cluster = 'hier', minClusterSize = 10)
unique(clusters_no_it$clusters$Cluster)

no.it.clusters <- unique(clusters_no_it$clusters$Cluster)
```



# Select paths related to Calcium and ATP:

```{r}
ca.paths <- clusters_no_it$clusters[clusters_no_it$clusters$Cluster %in% "regulation of sequestering of calcium ion",]$Pathway

atp.paths <- clusters_no_it$clusters[clusters_no_it$clusters$Cluster %in% "mitochondrial respirasome",]$Pathway
```


```{r}
#set.seed(4)
set.seed(4)

tiff("C://Users/Gerard/Desktop/PathCluster_Abstract_4.tiff", height = 2500, width = 2500, res = 350)
plotPathClusters(
  enrichment = go_enrich_no_it@result,
  sim = clusters_no_it$similarity,
  clusters = clusters_no_it$clusters,
  fontSize = 10,
  outerCutoff = 0.1,
  drawEllipses = TRUE,
  nodeSize = "Count",
  colorBy = "pvalue",
  colorType = 'pval',
  repelLabels = TRUE
)
dev.off()
```



# ATP pathways:
```{r}
df <- go_enrich_no_it@result

df <- df[df$Description %in% atp.paths,]

df_split <- df %>%
 separate_rows(geneID, sep = "/")

gene_counts <- table(df_split$geneID)
gene_counts <- sort(gene_counts, decreasing = TRUE)

# Genes present in all the ATP pathways (17):
genes <- head(gene_counts, 22)

# Genes present in the complexes that form the electron transport chain:
genes <- names(gene_counts)[grep("COX|NDUF|UQCC", names(gene_counts))]

all.genes <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/SigRes/Cases_Controls_Confounders.xlsx")

# Determine the directions of the genes in ATP pathways:
sum(all.genes[all.genes$Gene.Name %in% (genes),]$Beta < 0) # 21
sum(all.genes[all.genes$Gene.Name %in% (genes),]$Beta > 0) # 3

# Determine the directions of the genes in ATP pathways:
sum(all.genes[all.genes$Gene.Name %in% names(gene_counts),]$Beta < 0)
sum(all.genes[all.genes$Gene.Name %in% names(gene_counts),]$Beta > 0)


# Determine the novelty of the genes in ATP pathways:
new.genes <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/NewGenes.xlsx")
new.genes[new.genes$Gene.Name %in% names(genes),]
```



# Calcium pathways:

```{r}
df <- go_enrich_no_it@result

df <- df[df$Description %in% ca.paths,]

df_split <- df %>%
 separate_rows(geneID, sep = "/")

gene_counts <- table(df_split$geneID)
gene_counts <- sort(gene_counts, decreasing = TRUE)
names(gene_counts)

# Genes present in all the Ca pathways (17):
genes <- head(gene_counts, 12)
names(genes)

all.genes <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/AllGenes.xlsx")

# Determine the directions of the genes in Calcium pathways:
all.genes[all.genes$Gene.Name %in% names(genes),]
sum(all.genes[all.genes$Gene.Name %in% names(genes),]$Beta < 0)
sum(all.genes[all.genes$Gene.Name %in% names(genes),]$Beta > 0)

100*(82/(82+74))
100*(8/(8+4))

# Determine the novelty of the genes in Calcium pathways:
new.genes <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/NewGenes.xlsx")
new.genes[new.genes$Gene.Name %in% names(genes),]
```




### DEG New Genes:

# Load ischemic time differentially expressed genes and new genes:

```{r}
new <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/NewGenes.xlsx")
```


```{r}
Results_New <- Results[rownames(Results) %in% new$Gene.Name,] # 2,851
```


### Plot P-values:

```{r}
hist(Results_New$pvalue, main = "", xlab = "Pvalue")
```


```{r}
gene_list <- substr(Results_New[Results_New$fdr < 0.05,]$gene_id, 1, 15)

entrez_ids <- bitr(gene_list, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db)

# GO enrichment analysis:
go_enrich_new <- enrichGO(gene    = entrez_ids$ENTREZID,
                      OrgDb         = org.Hs.eg.db,
                      keyType       = "ENTREZID",
                      ont           = "ALL", # Choose from "BP" (Biological Process), "CC" (Cellular Component), or "MF" (Molecular Function)
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff  = 0.2,
                      readable = T)
```


### Save results:

```{r}
#write.xlsx(go_enrich_new@result, "C://Users/Gerard/Desktop/AAA/RNAseq/Supplementary_Tables/GO_Enrich_New.xlsx")

go_enrich_new@result <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Supplementary_Tables/GO_Enrich_New.xlsx")
```

```{r}
go_enrich_new@result
```


```{r}
clusters_new <- findPathClusters(go_enrich_new@result, cluster = 'hier', minClusterSize = 10)
unique(clusters_new$clusters$Cluster)

no.it.clusters <- unique(clusters_new$clusters$Cluster)
```


```{r}
ribo.paths <- clusters_new$clusters[clusters_new$clusters$Cluster %in% "mitochondrial ribosome",]$Pathway
```


# Ribosome pathways:
```{r}
df <- go_enrich_new@result

df <- df[df$Description %in% ribo.paths,]

df_split <- df %>%
 separate_rows(geneID, sep = "/")

gene_counts <- table(df_split$geneID)
gene_counts <- sort(gene_counts, decreasing = TRUE)

# Genes present in all the "Mitochondrial ribosome" pathways (11):
genes <- head(gene_counts, 22)


all.genes <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/AllGenes.xlsx")

# Determine the directions of the genes in ATP pathways:
sum(all.genes[all.genes$Gene.Name %in% names(genes),]$Beta < 0)
sum(all.genes[all.genes$Gene.Name %in% names(genes),]$Beta > 0)


# Determine the novelty of the genes in "Mitochondrial ribosome" pathways:
new.genes <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/NewGenes.xlsx")
new.genes[new.genes$Gene.Name %in% names(genes),]
```









# Compare GO enrich results:

```{r}
intersect(go_enrich$Description, go_enrich_no_it$Description)

setdiff(go_enrich$Description, go_enrich_no_it$Description)


# Compare clusters:

intersect(it.clusters, no.it.clusters)

setdiff(it.clusters, no.it.clusters)
```


```{r}
# GO enrichment dot plot:
png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Enrichment/GO_Cases_vs_Controls_No_Smoking_No_IT.png", width = 4*800, height = 4*600, res = 300)
dotplot(go_enrich_no_it, showCategory = 20) + ggtitle("GO Enrichment") +
  theme(plot.title = element_text(hjust = 0.5))
dev.off()
```

```{r}
# KEGG enrichment analysis
kegg_enrich_no_IT <- enrichKEGG(gene   = entrez_ids$ENTREZID,
                          organism     = "hsa",
                          keyType      = "kegg",
                          pAdjustMethod = "BH",
                          pvalueCutoff = 0.05,
                          qvalueCutoff = 0.2)
```


### Save results:

```{r}
#write.xlsx(kegg_enrich_no_IT@result, "C://Users/Gerard/Desktop/AAA/RNAseq/Supplementary_Tables/KEGG_Enrich_NO_IT.xlsx")
```


```{r}
# KEGG enrichment dot plot
png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Enrichment/KEGG_Cases_vs_Controls_No_Smoking_No_IT.png", width = 4*800, height = 4*600, res = 300)
dotplot(kegg_enrich_no_IT, showCategory = 15) + ggtitle("KEGG Enrichment") +
  theme(plot.title = element_text(hjust = 0.5))
dev.off()
```



```{r}
edox <- setReadable(go_enrich_no_it, 'org.Hs.eg.db', 'ENTREZID')
p3 <- cnetplot(edox, foldChange=entrez_ids$ENTREZID, circular = TRUE, colorEdge = TRUE)
```


```{r}
png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Enrichment/CnetPlot_No_IT.png", height = 2500, width = 2500)
p3
dev.off()
```


```{r, warning=F, echo=F, message=F}
edox2 <- pairwise_termsim(edox)
p2 <- treeplot(edox2, showCategory = 200, nCluster = 5)
```

```{r}
png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Enrichment/TreePlot_No_IT.png", height = 2000, width = 2000)
p2
dev.off()
```


```{r}
edo <- pairwise_termsim(go_enrich)
p4 <- emapplot(edo, cex_category=1.5,layout="kk", showCategory = 200) 
```


```{r}
png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Enrichment/EnrichmentMap_No_IT.png", height = 2000, width = 2000)
p4
dev.off()
```











