---
title: "1.3.Manual_Analysis_RSEM_Cases"
author: "Gerard Temprano Sagrera"
date: "2023-02-24"
output: html_document
---

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(data.table);library(edgeR);library(tidyverse);library(openxlsx);library(preprocessCore);library("org.Hs.eg.db");library(stats);library(qvalue);library(ggpubr);library(VennDiagram);library(grid);library(clusterProfiler)
```

### Load Annotation file to change format names:

```{r}
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v26.annotation.fixed.gtf.gz")
```

### Counts are already in TPM from RSEM:

```{r}
counts.tpm <- read.table("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.RSEM.Cases.txt")
```

### Subset significant genes when comparing cases vs controls:

```{r}
#sig.genes <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Significant_genes.txt", header = F)
#sig.genes <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Significant_genes_smoking.txt", header = F)

#counts.tpm <- counts.tpm[rownames(counts.tpm) %in% sig.genes$V1,]
```

### Load technical data and subset to cases:

```{r}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/First Analysis/Technical_CompleteV2.xlsx")
tec <- tec[tec$SampleID %in% colnames(counts.tpm),]
```

### Load biological data and subset to cases:

```{r}
bio <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/20221115_Datos Pacientes.xlsx")
bio <- bio[bio$Muestra %in% colnames(counts.tpm),]

# Smoking:
bio$Smoking2 <- ifelse(bio$Smoking == 0, 0, 1)
```


### By Diameter:

```{r}
dge <- DGEList(counts = counts.tpm)
```

### Remove lowely expressed genes using manual filtering:

```{r}
keep <- apply(dge$counts, 1, function(x) sum(x > 0.5) >= 48)
dge <- dge[keep,,keep.lib.sizes=FALSE]

paste0("We keep ", sum(keep), " genes") # 14,779
paste0("We remove ", sum(!keep), " genes") # 12,511
```


### Apply quantile normalization to counts:

```{r}
counts.tpm <- as.data.frame(normalize.quantiles(as.matrix(dge$counts), keep.names = TRUE))
```

### Manual Differential Expression:

```{r}
print(paste0("There are ",sum(is.na(bio$aortc_diameter_mm))," NAs"))

Pvalue <-
  apply(counts.tpm, 1, function(x)
    summary(
      lm(
        x ~ bio$aortc_diameter_mm + tec$Date + tec$Batch + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$age + bio$SEXO
      )
    )[4][[1]][2, 4])
Coef <-
  apply(counts.tpm, 1, function(x)
    summary(
      lm(
        x ~ bio$aortc_diameter_mm + tec$Date + tec$Batch + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$age + bio$SEXO
      )
    )[4][[1]][2, 1])
```


### Plot P-values:

```{r}
png(file = "C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Results/Hist_Diameter.png", height = 10000, width = 10000, res = 1200)
hist(Pvalue, main = "")
dev.off()

hist(Pvalue, main = "")
```

### Create qvalue object:

```{r}
qobj <- qvalue(p = Pvalue)
```

```{r, warning=FALSE}
#summary(qobj)
#hist(qobj)
#plot(qobj)
```

```{r}
Results <- data.frame(
  pvalue = Pvalue,
  qvalue = qobj$qvalues,
  fdr = p.adjust(Pvalue, method = "fdr"),
  pi1 = 1-qobj$pi0,
  Beta = Coef
)

Results <- merge(Results, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
names(Results)[names(Results) == "Row.names"] <- "gene_id"
Results <- Results[!duplicated(Results$gene_name),]
Results <- Results %>% remove_rownames %>% column_to_rownames(var="gene_name")
Results <- Results[order(Results$pvalue),]

nrow(Results[Results$fdr < 0.05,]) # 32


# Load ischemic time results:

isc <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Ischemic/Ichemic_Time_Genes.xlsx")
head(isc)


length(intersect(rownames(Results[Results$fdr < 0.05,]), isc$Gene_Name)) # 19

#write.xlsx(Results, "C://Users/Gerard/Desktop/AAA/RNAseq/AllRes/Diameter.xlsx", quote = F, rowNames = T, colNames = T)

#write.xlsx(Results[Results$fdr < 0.05,], "C://Users/Gerard/Desktop/AAA/RNAseq/SigRes/Diameter.xlsx", sep = "\t", rowNames = T)
```


### Diameter:

```{r}
gene_list <- substr(Results[Results$fdr < 0.05,]$gene_id, 1, 15)

entrez_ids <- bitr(gene_list, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db)

# GO enrichment analysis
go_enrich <- enrichGO(gene          = entrez_ids$ENTREZID,
                      OrgDb         = org.Hs.eg.db,
                      keyType       = "ENTREZID",
                      ont           = "ALL",
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff  = 0.2,
                      minGSSize = 2)
```

```{r}
# GO enrichment dot plot:
#png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/GO_Enrichment_Diameter.png", width = 4*800, height = 4*600, res = 300)
#dotplot(go_enrich, showCategory = 10) + ggtitle("GO Enrichment")
#dev.off()
```

```{r}
# KEGG enrichment analysis
#kegg_enrich <- enrichKEGG(gene         = entrez_ids$ENTREZID,
#                          organism     = "hsa",
#                          keyType      = "kegg",
#                          pAdjustMethod = "BH",
#                          pvalueCutoff = 0.05,
#                          qvalueCutoff = 0.2)
```

```{r}
# KEGG enrichment dot plot
#png("C://Users/Gerard/Desktop/KEGG_Enrichment_Diameter.png", width = 4*800, height = 4*600, res = 300)
#dotplot(kegg_enrich, showCategory = 15) + ggtitle("KEGG Enrichment")
#dev.off()
```



### Volcano Plot:

```{r}
#significant_genes <- Results %>% filter(qvalue < 0.05)

#g<- ggplot(data = Results, aes(x = Beta, y = -log(pvalue))) +
#  geom_point(aes(color = qvalue < 0.05), size = 0.5) +
#  scale_color_manual(values = c("gray", "red")) +
#  geom_hline(yintercept = -log10(0.05)) +
#  geom_vline(xintercept = 1) +
#  geom_vline(xintercept = -1) +
#  xlim(-20,20) +
#  theme_minimal() +
#  geom_text(data = significant_genes,
#            aes(label = rownames(significant_genes), hjust = 0, vjust = -0.5),
#            size = 3, check_overlap = TRUE)

#png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Results/Volcano_Diameter.png", width = 4*800, height = 4*600, res = 300)
#g
#dev.off()
```



### Compare with previous results (Tunica-Specific) and Cases vs Controls results:

```{r}
# Load Results:
cc <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/SigRes/Cases_Controls.xlsx")
dia <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/SigRes/Diameter.xlsx")

med <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/TunicaSpecificResults.xlsx", sheet = 1)
med$Gene.Symbol <- gsub("-", "", med$Gene.Symbol)

adv <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/TunicaSpecificResults.xlsx", sheet = 2)
adv$Gene.Symbol <- gsub("-", "", adv$Gene.Symbol)
```


### By Type:

```{r}
counts.tpm <- read.table("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.RSEM.Cases.txt")
```


```{r}
dge <- DGEList(counts = counts.tpm)
```


### Remove lowely expressed genes using LIMMA:

```{r}
keep <- apply(dge$counts, 1, function(x) sum(x > 0.5) >= 48)
dge <- dge[keep,,keep.lib.sizes=FALSE]

paste0("We keep ", sum(keep), " genes") # 14,779
paste0("We remove ", sum(!keep), " genes") # 12,511
```


### Apply quantile normalization to counts:

```{r}
counts.tpm <- as.data.frame(normalize.quantiles(as.matrix(dge$counts), keep.names = TRUE))
```

```{r}
print(paste0("There are ", sum(is.na(
  bio$TIPOANEURISMA
)), " NAs"))

Pvalue <-
  apply(dge$counts, 1, function(x)
    summary(
      lm(
        x ~ bio$TIPOANEURISMA + tec$Date + tec$Batch + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$age + bio$SEXO
      )
    )[4][[1]][2, 4])
Coef <-
  apply(dge$counts, 1, function(x)
    summary(
      lm(
        x ~ bio$TIPOANEURISMA + tec$Date + tec$Batch + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$age + bio$SEXO
      )
    )[4][[1]][2, 1])
```


### Plot P-values:

```{r}
png(file = "C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Results/Hist_Type.png", height = 10000, width = 10000, res = 1200)
hist(Pvalue, main = "")
dev.off()

hist(Pvalue, main = "")
```

### Create qvalue object:

```{r}
qobj <- qvalue(p = Pvalue)
```

```{r, warning=FALSE}
#summary(qobj)
#hist(qobj)
#plot(qobj)
```

```{r}
Results <- data.frame(
  pvalue = Pvalue,
  qvalue = qobj$qvalues,
  fdr = p.adjust(Pvalue, method = "fdr"),
  pi1 = 1-qobj$pi0,
  Beta = Coef
)

Results <- merge(Results, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
names(Results)[names(Results) == "Row.names"] <- "gene_id"
Results <- Results[!duplicated(Results$gene_name),]
Results <- Results %>% remove_rownames %>% column_to_rownames(var="gene_name")
Results <- Results[order(Results$pvalue),]

nrow(Results[Results$fdr < 0.05,]) # 1

#write.xlsx(Results, "C://Users/Gerard/Desktop/AAA/RNAseq/AllRes/Type.xlsx", quote = F, rowNames = T, colNames = T)

#write.xlsx(Results[Results$fdr < 0.05,], "C://Users/Gerard/Desktop/AAA/RNAseq/SigRes/Type.xlsx", sep = "\t", rowNames = T)
```




### By Symptoms:

```{r}
counts.tpm <- read.table("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.RSEM.Cases.txt")
```


```{r}
dge <- DGEList(counts = counts.tpm)
```

### Remove lowely expressed genes using LIMMA:

```{r}
keep <- apply(dge$counts, 1, function(x) sum(x > 0.5) >= 48)
dge <- dge[keep,,keep.lib.sizes=FALSE]

paste0("We keep ", sum(keep), " genes") # 14,779
paste0("We remove ", sum(!keep), " genes") # 12,511
```

### Apply quantile normalization to counts:

```{r}
counts.tpm <- as.data.frame(normalize.quantiles(as.matrix(dge$counts), keep.names = TRUE))
```

```{r}
print(paste0("There are ", sum(is.na(
  bio$symptomatic
)), " NAs"))

Pvalue <-
  apply(dge$counts, 1, function(x)
    summary(
      lm(
        x ~ bio$symptomatic + tec$Date + tec$Batch + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$age + bio$SEXO
      )
    )[4][[1]][2, 4])
Coef <-
  apply(dge$counts, 1, function(x)
    summary(
      lm(
        x ~ bio$symptomatic + tec$Date + tec$Batch + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$age + bio$SEXO
      )
    )[4][[1]][2, 1])
```


### Plot P-values:

```{r}
png(file = "C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Results/Hist_Symptoms.png", height = 10000, width = 10000, res = 1200)
hist(Pvalue, main = "")
dev.off()

hist(Pvalue, main = "")
```

### Create qvalue object:

```{r}
qobj <- qvalue(p = Pvalue)
```

```{r, warning=FALSE}
#summary(qobj)
#hist(qobj)
#plot(qobj)
```

```{r}
Results <- data.frame(
  pvalue = Pvalue,
  qvalue = qobj$qvalues,
  fdr = p.adjust(Pvalue, method = "fdr"),
  pi1 = 1-qobj$pi0,
  Beta = Coef
)

Results <- merge(Results, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
names(Results)[names(Results) == "Row.names"] <- "gene_id"
Results <- Results[!duplicated(Results$gene_name),]
Results <- Results %>% remove_rownames %>% column_to_rownames(var="gene_name")
Results <- Results[order(Results$pvalue),]

nrow(Results[Results$fdr < 0.05,]) # 30

#write.xlsx(Results, "C://Users/Gerard/Desktop/AAA/RNAseq/AllRes/Symptoms.xlsx", quote = F, rowNames = T, colNames = T)

#write.xlsx(Results[Results$fdr < 0.05,], "C://Users/Gerard/Desktop/AAA/RNAseq/SigRes/Symptoms.xlsx", sep = "\t", rowNames = T)
```

### Symptoms:

```{r}
gene_list <- substr(Results[Results$fdr < 0.05,]$gene_id, 1, 15)

entrez_ids <- bitr(gene_list, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db)

# GO enrichment analysis
go_enrich <- enrichGO(gene          = entrez_ids$ENTREZID,
                      OrgDb         = org.Hs.eg.db,
                      keyType       = "ENTREZID",
                      ont           = "ALL", # Choose from "BP" (Biological Process), "CC" (Cellular Component), or "MF" (Molecular Function)
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff  = 0.2,
                      readable = T)
```

```{r}
# GO enrichment dot plot:
#png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Enrichment/GO_Enrichment_Symptoms.png", width = #4*800, height = 4*600, res = 300)
#dotplot(go_enrich, showCategory = 10) + ggtitle("GO Enrichment")
#dev.off()
```


### Save results:

```{r}
write.xlsx(go_enrich@result, "C://Users/Gerard/Desktop/AAA/RNAseq/Supplementary_Tables/GO_Enrich_Symptoms.xlsx")
```


```{r}
# KEGG enrichment analysis
#kegg_enrich <- enrichKEGG(gene         = entrez_ids$ENTREZID,
#                          organism     = "hsa",
#                          keyType      = "kegg",
#                          pAdjustMethod = "BH",
#                          pvalueCutoff = 0.05,
#                          qvalueCutoff = 0.2)
```


```{r}
# KEGG enrichment dot plot
#png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Enrichment/KEGG_Symptoms.png", width = 4*800, height = 4*600, res = 300)
#dotplot(kegg_enrich, showCategory = 15) + ggtitle("KEGG Enrichment")
#dev.off()
```


### Volcano Plot:

```{r}
#significant_genes <- Results %>% filter(fdr < 0.05)

#g<- ggplot(data = Results, aes(x = Beta, y = -log(pvalue))) +
#  geom_point(aes(color = qvalue < 0.05), size = 0.5) +
#  scale_color_manual(values = c("gray", "red")) +
#  geom_hline(yintercept = -log10(0.05)) +
#  geom_vline(xintercept = 1) +
#  geom_vline(xintercept = -1) +
#  xlim(-20,20) +
#  theme_minimal() +
#  geom_text(data = significant_genes,
#            aes(label = rownames(significant_genes), hjust = 0, vjust = -0.5),
#            size = 3, check_overlap = TRUE)

#png("C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Results/Volcano_Type.png", width = 4*800, height = 4*600, res = 300)
#g
#dev.off()
```














