---
title: "1.3.1.Manual_Analysis_RSEM_Smoking"
author: "Gerard Temprano Sagrera"
date: "2023-04-27"
output: html_document
---

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(data.table);library(edgeR);library(tidyverse);library(openxlsx);library(preprocessCore);library("org.Hs.eg.db");library(stats);library(qvalue);library(ggpubr)
```

### Load Annotation file to change format names:

```{r}
ann <- fread("C://Users/Gerard/Desktop/AAA/RNAseq/Annotation/gencode.v26.annotation.fixed.gtf.gz")
```

### Convert TPM counts to DGElist object:

```{r}
counts.tpm <- read.table("C://Users/Gerard/Desktop/AAA/RNAseq/Gene_counts/Counts.RSEM.All.txt")
```

### Load technical data and subset to cases:

```{r}
tec <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/Tables/Technical_Complete.xlsx")
tec <- tec[tec$Seq_ID %in% colnames(counts.tpm),]
```

### Load biological data and subset to cases:

```{r}
bio <- read.xlsx("C://Users/Gerard/Desktop/AAA/RNAseq/20221115_Datos Pacientes.xlsx")
bio <- bio[bio$Muestra %in% colnames(counts.tpm),]

# Smoking:
bio$Smoking2 <- ifelse(bio$Smoking == 0, 0, 1)
```

### Convert TPM counts to DGElist object:

```{r}
dge <- DGEList(counts = counts.tpm)

# Set "Control" to be the first level:
AAA <- factor(x = c(rep("Case", 96), rep("Control", 44)),
              levels = c("Control", "Case"))
```

### Remove lowely expressed genes using manual filtering:

```{r}
keep <- apply(dge$counts, 1, function(x) sum(x > 0.5) >= 70)
dge <- dge[keep,,keep.lib.sizes=FALSE]

paste0("We keep ", sum(keep), " genes") # 14,675
paste0("We remove ", sum(!keep), " genes") # 12,615
```

### Apply quantile normalization to counts:

```{r}
counts.tpm <- as.data.frame(normalize.quantiles(as.matrix(dge$counts), keep.names = T))
```

### Manual Differential Expression:

```{r}
Pvalue <-
  apply(counts.tpm, 1, function(x)
    summary(
      lm(
        x ~ tec$Case + tec$FC + tec$Lane + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$age + bio$SEXO + bio$Smoking
      )
    )[4][[1]][2, 4])

Coef <-
  apply(counts.tpm, 1, function(x)
    summary(
      lm(
        x ~ tec$Case + tec$FC + tec$Lane + tec$GC_Mean + tec$RIN + tec$DV200 + tec$Qubit + bio$age + bio$SEXO + bio$Smoking
      )
    )[4][[1]][2, 1])
```


### Plot P-values:

```{r}
#png(file = "C://Users/Gerard/Desktop/AAA/RNAseq/Plots/Hist_Cases_vs_Controls_Smoking.png", height = 10000, width = 10000, res = 1200)
#hist(Pvalue, main = "", cex.main = 2)
#dev.off()

hist(Pvalue, cex.main = 2)
#hist(Pvalue_Current, main = "Current Smokers", cex.main = 2)
#hist(Pvalue_Smoking, main = "Ever Smokers vs Never Smokers", cex.main = 2)
```

### Create qvalue object:

```{r}
qobj <- qvalue(p = Pvalue)
#qobj_Current <- qvalue(p = Pvalue_Current)
#qobj_Smoking <- qvalue(p = Pvalue_Smoking)
```


### Results Ever Smokers vs Never Smokers:

```{r}
Results_Smoking <- data.frame(
  pvalue = Pvalue,
  qvalue = qobj$qvalues,
  fdr = p.adjust(Pvalue, method = "fdr"),
  pi1 = 1-qobj$pi0,
  Beta = Coef
)

Results_Smoking <- merge(Results_Smoking, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
names(Results_Smoking)[names(Results_Smoking) == "Row.names"] <- "gene_id"
Results_Smoking <- Results_Smoking[!duplicated(Results_Smoking$gene_name),]
Results_Smoking <- Results_Smoking %>% remove_rownames %>% column_to_rownames(var="gene_name")
Results_Smoking <- Results_Smoking[order(Results_Smoking$pvalue),]

nrow(Results_Smoking[Results_Smoking$fdr < 0.05,]) # 6372

#fwrite(Results_Smoking[Results_Smoking$fdr < 0.05,], "C://Users/Gerard/Desktop/AAA/RNAseq/SigRes/Smoking.txt", sep = "\t", row.names = T)
```


### Results Old Smokers (1) vs Never Smokers (0):

```{r}
Results_Past <- data.frame(
  pvalue = Pvalue_Past,
  qvalue = qobj_Past$qvalues,
  fdr = p.adjust(Pvalue_Past, method = "fdr"),
  pi1 = 1-qobj_Past$pi0,
  Beta = Coef_Past
)

Results_Past <- merge(Results_Past, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
Results_Past <- Results_Past[,-1]
Results_Past <- Results_Past[!duplicated(Results_Past$gene_name),]
Results_Past <- Results_Past %>% remove_rownames %>% column_to_rownames(var="gene_name")
Results_Past <- Results_Past[order(Results_Past$pvalue),]

nrow(Results_Past[Results_Past$fdr < 0.05,]) # 1.

#fwrite(Results_Past[Results_Past$fdr < 0.05,], "C://Users/Gerard/Desktop/AAA/RNAseq/SigRes/Smoking_Past.txt", sep = "\t", row.names = T)
```


### Results Current Smokers (2) vs Never Smokers (0):

```{r}
Results_Current <- data.frame(
  pvalue = Pvalue_Current,
  qvalue = qobj_Current$qvalues,
  fdr = p.adjust(Pvalue_Current, method = "fdr"),
  pi1 = 1-qobj_Current$pi0,
  Beta = Coef_Current
)

Results_Current <- merge(Results_Current, ann[,c("gene_id","gene_name")], by.x = 0, by.y = "gene_id")
Results_Current <- Results_Current[,-1]
Results_Current <- Results_Current[!duplicated(Results_Current$gene_name),]
Results_Current <- Results_Current %>% remove_rownames %>% column_to_rownames(var="gene_name")
Results_Current <- Results_Current[order(Results_Current$pvalue),]

nrow(Results_Current[Results_Current$fdr < 0.05,]) # 0.

#fwrite(Results_Current[Results_Current$fdr < 0.05,], "C://Users/Gerard/Desktop/AAA/RNAseq/SigRes/Smoking_Current.txt", sep = "\t", row.names = T)
```


### DEG:

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(clusterProfiler);library(org.Hs.eg.db);library(enrichplot);library(ggplot2);library(enrichplot);library("ggupset");library(ggtree);library(aPEAR)
```


```{r}
gene_list <- substr(Results_Smoking[Results_Smoking$fdr < 0.05,]$gene_id, 1, 15)

entrez_ids <- bitr(gene_list, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db)

# GO enrichment analysis:
go_enrich <- enrichGO(gene          = entrez_ids$ENTREZID,
                      OrgDb         = org.Hs.eg.db,
                      keyType       = "ENTREZID",
                      ont           = "ALL", # Choose from "BP" (Biological Process), "CC" (Cellular Component), or "MF" (Molecular Function)
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff  = 0.2,
                      readable = T)
```



### Save results:

```{r}
#write.xlsx(go_enrich@result, "C://Users/Gerard/Desktop/AAA/RNAseq/Supplementary_Tables/GO_Enrich.xlsx")
```




```{r}
clusters <- findPathClusters(go_enrich@result, cluster = 'hier', minClusterSize = 15)
clusters$clusters
smoking.clusters <- unique(clusters$clusters$Cluster)
```








